A Practical Guide For SQL Tuning (#19108)
pingcap/docs: sql-tuning-best-practice.md

Reviewer: @dbsid Please ensure there is only one first-level heading (H1) per document, which should match the title `A Practical Guide For SQL Tuning` specified in the frontmatter. Currently, there are five H1 headings. Could you restructure the document to use appropriate heading levels (H2, H3, etc.) for the other sections?

Reviewer: Preview: https://pingcap-docsite-preview-git-preview-pi-ab8d5e-oreoxmts-projects.vercel.app/tidb/dev/sql-tuning-best-practice/

Examples:

- Creating an index for frequently queried columns can significantly reduce resource usage, especially for queries that access a small percentage of table data.
- Utilizing index-only scans for queries that return a limited number of sorted results can avoid full table scans and sorting operations.

## Balance Workload Distribution

In a distributed architecture like TiDB, it's essential to maintain a balanced workload across multiple TiKV nodes. This is crucial for achieving optimal performance. For guidance on identifying and resolving read and write hotspots, please refer to the [document](https://docs.pingcap.com/tidb/stable/troubleshoot-hot-spot-issues#optimization-of-small-table-hotspots). By implementing these strategies, you can ensure that your TiDB cluster efficiently utilizes all available resources and avoids bottlenecks caused by uneven workload distribution or serialization on individual TiKV nodes.

# Identifying High-Load SQL
The most efficient way to identify Resource-Intensive SQL is using TiDB Dashboard, There are other tools like views and logs available as well.

## Monitoring SQL Statements by Using TiDB Dashboard
### SQL Statements Panel 

> In TiDB Dashboard, navigate to SQL Statements panel, which helps us identify the following:

Reviewer 2: ```suggestion
In [TiDB Dashboard](dashboard/dashboard-overview.md), navigate to SQL Statements panel, which helps us identify the following:
```

1. The SQL statement with the highest total latency is the one that takes the longest time to execute out of multiple executions of the same SQL statement. 
2. It can also display the number of times each SQL statement has been executed cumulatively, allowing us to find the SQL statement with the highest execution frequency.

> 3. Clicking on each SQL statement allows us to delve deeper into the EXPLAIN ANALYZE results.

Reviewer 2: ```suggestion
3. Clicking on each SQL statement allows us to delve deeper into the `EXPLAIN ANALYZE` results.
```

SQL statements are normalized as templates, where literals and bind variables are replaced by '?'. This normalization and sorting process allows you to quickly pinpoint the most resource-intensive queries that may require optimization.
![sql-statements-default](/media/sql-tuning/sql-statements-default.png)

### Slow Queries Panel Default Display
In the TiDB Dashboard, we can find the Slow Query panel, which displays all SQL statements whose execution time exceeds the time threshold set by the system variable tidb_slow_log_threshold (default 300 milliseconds). 

On Slow Queries panel, we can find:

1. The slowest SQL queries.
2. The SQL query that reads the most data from TiKV.

> 3. The EXPLAIN ANALYZE output from drilling down the SQL statement by clicking it.

Reviewer 2: ```suggestion
3. The `EXPLAIN ANALYZE` output from drilling down the SQL statement by clicking it.
```

4. Please note that on the Slow Queries panel, we cannot get the frequency of the SQL statement execution. Once the execution elapsed time exceeds tidb_slow_log_threshold for single instance, the query is then listed on the Slow Queries panel.
![slow-query-default](/media/sql-tuning/slow-query-default.png)

## Other Tools for Identifying Top SQL

In addition to TiDB Dashboard, there are several other tools available to identify resource-intensive SQL queries:

Each tool offers unique insights and can be valuable for different analysis scenarios. Using a combination of these tools allows for comprehensive SQL performance monitoring and optimization.
- [slow query log](https://docs.pingcap.com/tidb/stable/identify-slow-queries)
- [statements_summary view](https://docs.pingcap.com/tidb/stable/statement-summary-tables#statements_summary)
- [top sql feature](https://docs.pingcap.com/tidb/stable/top-sql)
- [expensive queries in tidb log](https://docs.pingcap.com/tidb/stable/identify-expensive-queries)
- [cluster_processlist view](https://docs.pingcap.com/tidb/stable/information-schema-processlist#cluster_processlist)

## Gathering Data on the SQL Identified

> For the top SQL statements identified, you can use [PLAN REPLAYER](https://docs.pingcap.com/tidb/stable/sql-plan-replayer) to capture and save the on-site information of a TiDB cluster. This tool allows you to recreate the execution environment for further analysis. The syntax for exporting SQL information is as follows:

Reviewer 2: This and many other links seem to link with the full URL of the doc instead of directly to the right doc. This means that people might jump between TiDB versions.

```suggestion
For the top SQL statements identified, you can use [PLAN REPLAYER](/sql-plan-replayer.md) to capture and save the on-site information of a TiDB cluster. This tool allows you to recreate the execution environment for further analysis. The syntax for exporting SQL information is as follows:
```

> ```SQL

Reviewer 2: ```suggestion
```sql
```

PLAN REPLAYER DUMP EXPLAIN [ANALYZE] [WITH STATS AS OF TIMESTAMP expression] sql-statement;
```

> Use EXPLAIN ANALYZE whenever possible, as it captures actual execution information in addition to the execution plan. This provides more accurate insights into query performance.

Reviewer 2: ```suggestion
Use `EXPLAIN ANALYZE` whenever possible, as it captures actual execution information in addition to the execution plan. This provides more accurate insights into query performance.
```

# SQL Tuning Guide

This guide focuses on providing actionable advice for beginners looking to optimize their SQL queries in TiDB. By following these best practices, you can ensure better query performance and SQL Tuning. We'll cover below topic:

> - Query Processing Workflow

Reviewer 2: ```suggestion

- Query Processing Workflow
```

- Optimizer Fundamentals
- Statistics Management
- How TiDB build A Execution Plan
- Understand Execution Plan
- Index Strategy in TiDB
  - SQL Tuning with a Covered Index
  - SQL Tuning with a Composite Index Involing Sorting
  - SQL Tuning with a Composite Index for Efficient Filtering and Sorting
  - Composite Index Strategy Guidelines
  - The Cost of Indexing

## Query Processing Workflow

> the client sends a SQL statement to the protocol layer of TiDB server. The protocol layer is responsible for handling the connection between TiDB Server and the client, receiving SQL statement from the client, and returning data to the client.

Reviewer 2: ```suggestion

The client sends a SQL statement to the protocol layer of TiDB server. The protocol layer is responsible for handling the connection between TiDB Server and the client, receiving SQL statement from the client, and returning data to the client.
```

- SQL
- Schema Design
- Statistics
  - Table
  - Index
  - Column

Based on the input, The cost model will produce the execution plan, which includes the details how the system execute the sql, including 
- Access Method
- Join Method
- Join Order

The optimizer is as good as the information it receives. Therefore, ensuring up-to-date statistics and well-designed indexes is critical.

## Statistics Management

Statistics are essential to the TiDB optimizer. TiDB uses statistics as input to the optimizer to estimate the number of rows processed in each plan step for a SQL statement.

statistics is generally divided into two levels: table level and column level. 
- For table-level statistics, it includes the total number of rows in the table and the number of rows that have been modified since the last collection of statistics. 
- The column-level statistics information is more abundant, including histograms, Count-Min Sketch, Top-N (values or indexes with the highest occurrences), distribution and quantity of different values, and the number of null values, and so on.

To ensure the statistics are healthy and representative, you can use the following commands:

> 1. `SHOW STATS_META`: This command provides metadata about table statistics.
> 2. `SHOW STATS_HEALTHY`: This command shows the health status of table statistics.

Reviewer 2: ```suggestion
1. [`SHOW STATS_META`](sql-statements/sql-statement-show-stats-meta.md): This command provides metadata about table statistics.
2. [`SHOW STATS_HEALTHY`](sql-statements/sql-statement-show-stats-healthy.md): This command shows the health status of table statistics.
```

For example, you can use:

> ```SQL
> tidb> SHOW STATS_META WHERE table_name='T2'\G;
> *************************** 1. row ***************************
>  Db_name: test
>  Table_name: T2
> Partition_name:
>  Update_time: 2023-05-11 02:16:50
>  Modify_count: 20000
>  Row_count: 20000
> 1 row in set (0.03 sec)
> 
> tidb> SHOW STATS_HEALTHY WHERE table_name='T2'\G;
> *************************** 1. row ***************************
>     Db_name: test
>   Table_name: T2
> Partition_name:
>     Healthy: 0
> 1 row in set (0.00 sec)
> 
> ```

Reviewer 2: ````suggestion
```sql
SHOW STATS_META WHERE table_name='T2'\G;
```

```
*************************** 1. row ***************************
 Db_name: test
 Table_name: T2
Partition_name:
 Update_time: 2023-05-11 02:16:50
 Modify_count: 20000
 Row_count: 20000
1 row in set (0.03 sec)
```

```sql
SHOW STATS_HEALTHY WHERE table_name='T2'\G;
```

```
*************************** 1. row ***************************
    Db_name: test
  Table_name: T2
Partition_name:
    Healthy: 0
1 row in set (0.00 sec)
```
````

The output isn't valid SQL

In TiDB database, there are two ways to collect statistics: automatic collection and manual collection. In most case, the auto collection job works fine. Automatic collection is actually triggered when certain conditions are met for a table, and TiDB will automatically collect statistics. We commonly use three triggering conditions, which are: ratio, start_time and end_time.
tidb_auto_analyze_ratio: The healthiness trigger

> tidb_auto_analyze_start_time and tidb_auto_analyze_end_time: The allowed job window

Reviewer 2: ```suggestion
[`tidb_auto_analyze_start_time`](/system-variables.md#tidb_auto_analyze_start_time) and tidb_auto_analyze_end_time: The allowed job window
```

Same for many other sysvars

> ```SQL

Reviewer 2: ```suggestion

```sql
```

> mysql> show variables like '%auto_analyze%';

Reviewer 2: ```suggestion
SHOW VARIABLES LIKE 'tidb\_auto\_analyze%';
```

+-----------------------------------------+-------------+
| tidb_auto_analyze_ratio                 | 0.5         |
| tidb_auto_analyze_start_time            | 00:00 +0000 |
| tidb_auto_analyze_end_time              | 23:59 +0000 |
+-----------------------------------------+-------------+
```

In cases where automatic collection doesn't meet your needs, you can manually collect statistics using the `ANALYZE TABLE table_name` statement. This allows you to:

1. Adjust the sample rate for more accurate or faster analysis
2. Increase the number of top-N values collected
3. Gather statistics for specific columns only

It's important to note that after manual collection, subsequent automatic gathering jobs will inherit the new parameters. This means that any customizations you've made during manual collection will be carried forward in future automatic analyses.

Another common scenario is locking table statistics. This is useful when:
1. The statistics on the table are already representative of the data.
2. The table is very large and statistics collection is time-consuming.
3. You want to maintain statistics only during specific time windows.

> To lock the statistics for a table, you can use the following command `LOCK STATS table_name`.

Reviewer 2: ```suggestion
To lock the statistics for a table, you can use the following statement [`LOCK STATS table_name`](/sql-statements/sql-statement-lock-stats.md).
```

for more detail about statistics, please refer to [statistics](https://docs.pingcap.com/tidb/stable/statistics).

## How TiDB build A Execution Plan
An SQL statement undergoes optimization primarily in the optimizer through three stages:

> - Pre-Processing

Reviewer 2: ```suggestion

- Pre-Processing
```

- Logical Transformation
- Cost-based Optimization

### Pre-Processing

> The main actions in the pre-processing stage it to determine if the SQL statement can be executed by using Point_Get or Batch_Point_Get.

Reviewer 2: ```suggestion

The main actions in the pre-processing stage it to determine if the SQL statement can be executed by using Point_Get or Batch_Point_Get.
```

> Point_Get or Batch_Point_Get is to get 1 or 0 or many row only by using the TiKV key, the explicit or implicit (_tidb_rowid) primary key. For example, when id column is the primary key of a clustered index table, Point_Get is used to get the particular row. If a plan is identified as Point_Get, optimizer will skip the logical transformation and cost-based optimization.

Reviewer 2: ```suggestion
Point_Get or Batch_Point_Get is to get 1 or 0 or many row only by using the TiKV key, the explicit or implicit (`_tidb_rowid`) primary key. For example, when id column is the primary key of a clustered index table, Point_Get is used to get the particular row. If a plan is identified as Point_Get, optimizer will skip the logical transformation and cost-based optimization.
```

> ```SQL

Reviewer 2: ```suggestion
```sql
```

More Detail for Logical Transformation: https://docs.pingcap.com/tidb/stable/sql-logical-optimization.

### Cost-Based Optimization

TiDB uses statistics as input to the optimizer to estimate the number of rows processed in each plan step for a SQL statement, and associates a cost with each plan step. The Cost-Based Optimization estimates the cost of each available plan choice, including index accesses and the sequence of table joins, and produces a cost for each available plan. The optimizer then picks the execution plan with the lowest overall cost.

The below figure illustrates the various data access paths and row set operations that cost-based optimization can consider to develop the optimal execution plan. Furthermore, during the logical transformation stage, the query has already been rewritten for predicate push-down. In the cost-based optimization stage, TiKV expression push-down is further implemented when possible at TiKV layer. 

> Furthermore, confirming the algorithm for certain SQL operations, such as aggregation, join, and sorting, is essential. For instance, the aggregation operator may utilize either HASH_AGG or STREAM_AGG, while the join operator can select from HASH JOIN, MERGE JOIN, or INDEX JOIN. Likewise, various options are available for the sorting operator.

Reviewer 2: ```suggestion
Furthermore, confirming the algorithm for certain SQL operations, such as aggregation, join, and sorting, is essential. For instance, the aggregation operator may utilize either `HASH_AGG` or `STREAM_AGG`, while the join operator can select from `HASH JOIN`, `MERGE JOIN`, or `INDEX JOIN`. Likewise, various options are available for the sorting operator.
```

![cost-based-optimization](/media/sql-tuning/cost-based-optimization.png)

## Understanding Execution Plans

The execution plan represents the steps TiDB will follow to execute a SQL query. In this section, we will learn how to display and read the execution plan.

### Generating and Displaying Execution Plans

Beside access the execution plan information through TiDB Dashboard, TiDB provides a `EXPLAIN` statement to display the execution plan for a SQL query. Here's an example of using `EXPLAIN`:
- id: Operator name and the step unique identifier
- estRows: Estimated number of rows from the particular step
- task: The executor of the step
- access object: The object where the row sources are located
- operator info: Extended information about the operator regarding the step

```SQL

> tidb> EXPLAIN SELECT count(*) FROM trips WHERE start_date BETWEEN '2017-07-01 00:00:00' AND '2017-07-01 23:59:59';

Reviewer 2: ```suggestion
EXPLAIN SELECT COUNT(*) FROM trips WHERE start_date BETWEEN '2017-07-01 00:00:00' AND '2017-07-01 23:59:59';
```

For this and other output:
- Set the language to `sql` instead of `SQL`
- Only the query is valid SQL, not the output
- The prompt should not be in there and if it is in there it should probably be `mysql> ` as that's what users would see without additional configuration.

| StreamAgg_20             | 1.00        | root         |                   | funcs:count(Column#13)->Column#11                                                                  |
| └─TableReader_21         | 1.00        | root         |                   | data:StreamAgg_9                                                                                   |
|   └─StreamAgg_9          | 1.00        | cop[tikv]    |                   | funcs:count(1)->Column#13                                                                          |
|     └─Selection_19       | 250.00      | cop[tikv]    |                   | ge(trips.start_date, 2017-07-01 00:00:00.000000), le(trips.start_date, 2017-07-01 23:59:59.000000) |
|       └─TableFullScan_18 | 10000.00    | cop[tikv]    | table:trips       | keep order:false, stats:pseudo                                                                     |
+--------------------------+-------------+--------------+-------------------+----------------------------------------------------------------------------------------------------+
5 rows in set (0.00 sec)
```

> Additional Information in EXPLAIN ANALYZE Output

Reviewer 2: ```suggestion
Additional Information in [EXPLAIN ANALYZE](sql-statements/sql-statement-explain-analyze.md) Output
```

> Different from EXPLAIN, EXPLAIN ANALYZE executes the corresponding SQL statement, records its runtime information, and returns the information together with the execution plan. There runtime information is crucial for debugging query execution.

Reviewer 2: ```suggestion
Different from `EXPLAIN`, `EXPLAIN ANALYZE` executes the corresponding SQL statement, records its runtime information, and returns the information together with the execution plan. There runtime information is crucial for debugging query execution.
```

Description
- actRows: Number of rows output by the operator.
- execution info: Detailed execution information of the operator. time represents the total wall time from entering the operator to leaving the operator, including the total execution time of all sub-operators. If the operator is called many times by the parent operator then the time refers to the accumulated time. loops is the number of times the current operator is called by the parent operator.
- memory: Memory used by the operator.
- disk: Disk space used by the operator.

Note: Some attributes and explain table columns are omitted for improved formatting

> ```SQL
> tidb:db1> EXPLAIN ANALYZE SELECT SUM(pm.m_count)/COUNT(*) FROM
>     -> (SELECT COUNT(m.name) m_count
>     ->  FROM universe.moons m
>     ->  RIGHT JOIN
>     ->  (SELECT p.id, p.name
>     ->   FROM universe.planet_categories c
>     ->   JOIN universe.planets p
>     ->   ON c.id = p.category_id AND c.name = 'Jovian') pc
>     ->  ON m.planet_id = pc.id
>     ->  GROUP BY pc.name) pm;

Reviewer 2: ```suggestion
```sql
EXPLAIN ANALYZE
SELECT SUM(pm.m_count)/COUNT(*) FROM
(SELECT COUNT(m.name) m_count
FROM universe.moons m
RIGHT JOIN
(SELECT p.id, p.name
FROM universe.planet_categories c
JOIN universe.planets p
ON c.id = p.category_id AND c.name = 'Jovian') pc
ON m.planet_id = pc.id
GROUP BY pc.name) pm;
```

This makes the output copy-paste-able and makes it easier for a highlighting to work correctly

Let's apply the "first child first – recursive descent" rule to the first plan. When reading an execution plan, you should start from the top and work your down bottom. In the below example begin by looking at the `TableFullScan_18` (or the first child of the tree). In this case the access operator for table trips are implemented using full table scan. The rows produced by the tables scans will be consumed by the `Selection_19` operator. The `Selection_19` operator is to filter the data by `ge(trips.start_date, 2017-07-01 00:00:00.000000), le(trips.start_date, 2017-07-01 23:59:59.000000)`. Next the group-by operator `StreamAgg_9` is to implemented the aggregation `count(*)`. Be noted that the 3 operators `TableFullScan_18`, `Selection_19`, `StreamAgg_9` are pushdown to TiKV, which is marked as `cop[tikv]`, so that early filter and aggregation can be done in TiKV, to minize the data transfer between TiKV and TiDB. Finally the `TableReader_21` is to read the data from the `StreamAgg_9` operator, then finally the `StreamAgg_20` is to implemented the aggregation `count(*)`.

```SQL

> tidb> EXPLAIN SELECT count(*) FROM trips WHERE start_date BETWEEN '2017-07-01 00:00:00' AND '2017-07-01 23:59:59';

Reviewer 2: ```suggestion
EXPLAIN SELECT COUNT(*) FROM trips WHERE start_date BETWEEN '2017-07-01 00:00:00' AND '2017-07-01 23:59:59';
```

| StreamAgg_20             | 1.00        | root         |                   | funcs:count(Column#13)->Column#11                                                                  |
| └─TableReader_21         | 1.00        | root         |                   | data:StreamAgg_9                                                                                   |
|   └─StreamAgg_9          | 1.00        | cop[tikv]    |                   | funcs:count(1)->Column#13                                                                          |
|     └─Selection_19       | 250.00      | cop[tikv]    |                   | ge(trips.start_date, 2017-07-01 00:00:00.000000), le(trips.start_date, 2017-07-01 23:59:59.000000) |
|       └─TableFullScan_18 | 10000.00    | cop[tikv]    | table:trips       | keep order:false, stats:pseudo                                                                     |
+--------------------------+-------------+--------------+-------------------+----------------------------------------------------------------------------------------------------+
5 rows in set (0.00 sec)
```

> Let's apply the "first child first – recursive descent" rule to the second plan. In the below example begin from the top to bottom, by looking at the `IndexRangeScan_47` (the first child of the tree). For the table `stars`, the optimizer ony need to select the column `name` and `id`, the two columns can be met by the index `name(name)`. So for the table `star`, the root reader is `IndexReader_48`, rather than a `TableReader`.

Reviewer 2: ```suggestion

Let's apply the "first child first – recursive descent" rule to the second plan. In the below example begin from the top to bottom, by looking at the `IndexRangeScan_47` (the first child of the tree). For the table `stars`, the optimizer ony need to select the column `name` and `id`, the two columns can be met by the index `name(name)`. So for the table `star`, the root reader is `IndexReader_48`, rather than a `TableReader`.
```

In the execution plan below, the query ran for 5 minutes and 51 seconds before being canceled. Let's analyze the key issues:

1. Severe underestimation: The first child operator `IndexReader_76` reads data from the index `index_orders_on_adjustment_id(adjustment_id)`. The actual number of rows (`actRows`) is 256,811,189, which is drastically higher than the estimated 1 row (`estRows`).

2. Memory overflow: Due to this underestimation, the hash join operator `HashJoin_69` attempts to build a hash table with far more data than anticipated. This results in excessive memory usage (22.6GB) and disk usage (7.65GB).

3. Query termination: The `actRows` is 0 for `HashJoin_69` and subsequent operators, indicating that the hash join consumed too much memory, likely causing the query to be terminated by memory control mechanisms.

4. Incorrect join order: The root cause of this inefficient plan is the severe underestimation of `estRows` for `IndexRangeScan_75`, which led to an incorrect join order decision by the optimizer.

To address these issues, need to ensure that table statistics are up-to-date, especially for the `orders` table and the `index_orders_on_adjustment_id` index.

> ```SQL

Reviewer 2: ```suggestion
```
```

1. Accurate estimation: The `estRows` values are now much closer to the `actRows`, indicating that the statistics have been updated and are more accurate.
2. Efficient join order: The query now starts with a `TableReader` on the `labels` table, followed by an `IndexJoin` with the `rates` table, and finally another `IndexJoin` with the `orders` table. This join order is more efficient given the actual data distribution.
3. No memory overflow: Unlike the previous plan, there are no signs of excessive memory or disk usage, indicating that the query executes within expected resource limits.
4. Complete execution: All operators show non-zero `actRows`, confirming that the query completes successfully without being terminated due to resource constraints.

This optimized plan demonstrates the importance of accurate statistics and proper join order in query performance. The dramatic reduction in execution time (from 351 seconds to 1.96 seconds) highlights the potential impact of addressing estimation errors and choosing appropriate execution strategies.

> ```SQL

Reviewer 2: ```suggestion
```
```

|   └─TableReader_84(Probe)             | 984.56   | 1998    | root      |                                                                                        | time:207.6ms, ...| N/A      | N/A  |
|     └─Selection_83                    | 984.56   | 1998    | cop[tikv] |                                                                                        | tikv_task:{pro...| N/A      | N/A  |
|       └─TableRangeScan_82             | 1000.00  | 2048    | cop[tikv] | table:orders                                                                           | tikv_task:{pro...| N/A      | N/A  |
+---------------------------------------+----------+---------+-----------+----------------------------------------------------------------------------------------+---------------...+----------+------+
```

## Index Strategy in TiDB

TiDB is a distributed SQL database that completely decouples the SQL layer (TiDB) from the storage layer (TiKV). Unlike traditional databases, TiDB does not have a buffer pool to cache data at the compute node. As a result, the performance of SQL queries and the TiDB cluster is closely tied to the number of key-value (KV) requests that need to be processed.

> In TiDB, leveraging indexes effectively is crucial for performance tuning, as it can significantly reduce the number of KV RPC (Remote Procedure Call) requests. By minimizing these requests, you can greatly improve query performance and overall system efficiency. Here are some key strategies:

Reviewer 2: ```suggestion
In TiDB, leveraging indexes effectively is crucial for performance tuning, as it can significantly reduce the number of KV [RPC](/glossary.md#remote-procedure-call-rpc) requests. By minimizing these requests, you can greatly improve query performance and overall system efficiency. Here are some key strategies:
```

|   └─HashAgg_7(Probe)         | 1.00       | 588434.48  | 301     | cop[tikv] |                                                              | time:3m46.7s, loops:260, cop_task: {num: 301,           ...|
|     └─Selection_13           | 1271.37    | 561549.27  | 2566562 | cop[tikv] |                                                              | tikv_task:{proc max:10s, min:0s, avg: 915.3ms,          ...|
|       └─TableRowIDScan_12    | 1309.50    | 430861.31  | 2597411 | cop[tikv] | table:logs                                                   | tikv_task:{proc max:10s, min:0s, avg: 908.7ms,          ...|
+-------------------------------+------------+---------+-----------+--------------------------------------------------------------------------+------------------------------------------------------------+
```

After creating the covered index below, which includes the additional columns source_type, target_type, and amount, the query execution time improved to 90ms, and TiDB only needed to send a single cop task to TiKV for data scanning.

> ```SQL

Reviewer 2: ```suggestion
```sql
```

> Create index logs_covered on logs(snapshot_id, user_id, status, source_type, target_type, amount); 

Reviewer 2: ```suggestion
CREATE INDEX logs_covered ON logs(snapshot_id, user_id, status, source_type, target_type, amount); 
```

```

> ```SQL

Reviewer 2: ```suggestion
```
```

|   └─HashAgg_6                 | 1.00       | 1       | cop[tikv] |                                                                                                                                 | tikv_task:{time:88ms, loops:52},         ...|
|     └─Selection_12            | 5245632.33 | 52863   | cop[tikv] |                                                                                                                                 | tikv_task:{time:80ms, loops:52}          ...|
|       └─IndexRangeScan_11     | 5245632.33 | 52863   | cop[tikv] | table:logs, index:logs_covered(snapshot_id, user_id, status, source_type, target_type, amount)                                  | tikv_task:{time:60ms, loops:52}          ...|
+-------------------------------+------------+---------+-----------+---------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------+
```

### SQL Tuning with a Composite Index Involing Sorting

> When optimizing SQL queries, especially those that include an `ORDER BY` clause, it is beneficial to create a composite index that encompasses both the filtering and sorting columns. This approach allows the database engine to efficiently access the required data while maintaining the desired order.

Reviewer 2: ```suggestion

When optimizing SQL queries, especially those that include an `ORDER BY` clause, it is beneficial to create a composite index that encompasses both the filtering and sorting columns. This approach allows the database engine to efficiently access the required data while maintaining the desired order.
```

For instance, consider the following query that retrieves logs based on specific conditions. The execution plan shows a duration of 170ms. TiDB employs the `logs_index` to perform an `IndexRangeScan_20` with the filter `snapshot_id = 459840`. Subsequently, it retrieves all columns from the table, resulting in 5715 rows being streamed back to TiDB after the `IndexLookUp_23`, which then sorts the dataset and returns 1000 rows.
Note that the `id` column is the primary key, which means it is implicitly included in the `logs_idx` index. However, for `IndexRangeScan_20`, the order is not guaranteed because, after the index prefix column `snapshot_id`, there are two additional columns: `user_id` and `status`. Therefore, the ordering of `id` cannot be assured.

> ```SQL
> explain analyze SELECT  

Reviewer 2: ```suggestion
```sql
EXPLAIN ANALYZE SELECT  
```

logs.*
FROM
  logs
WHERE
  logs.snapshot_id = 459840
  AND logs.id > 998464
ORDER BY
  logs.id ASC
LIMIT
  1000
```

Original Plan

> ```SQL

Reviewer 2: ```suggestion

```
```

|   │ └─IndexRangeScan_20      | 433.47  | 7715    | cop[tikv] | table:logs, index:logs_idx(snapshot_id, user_id, status) | tikv_task:{proc max:4ms, min:4ms, avg: 4ms ...| range:[459840,459840], keep order:false    |
|   └─TableRowIDScan_21(Probe) | 19.98   | 5715    | cop[tikv] | table:logs                                               | time:301.6ms, loops:10, cop_task: {num: 3, ...| keep order:false                           | 
+------------------------------+---------+---------+-----------+----------------------------------------------------------+-----------------------------------------------+--------------------------------------------+
```

To optimize the query, we can create a new index on (snapshot_id, id) to ensure that for each snapshot_id value, the id is sorted in the index. By utilizing this new index, the query execution time is reduced to 96ms. Note that the keep order is true for `IndexRangeScan_33`, and the `TopN` is replaced with `Limit`. For the `IndexLookUp_35`, only 1000 rows are returned to TiDB, eliminating the need for additional sorting operations.

> ```SQL
> create index logs_new on logs(snapshot_id, id);
> ```

Reviewer 2: ````suggestion
```sql
CREATE INDEX logs_new ON logs(snapshot_id, id);
```
````

New Plan

> ```SQL

Reviewer 2: ````suggestion

```
````

+----------------------------------+---------+---------+-----------+----------------------------------------------+----------------------------------------------+----------------------------------------------------+
```

### SQL Tuning with Composite Indexes for Efficient Filtering and Sorting

The original query took 11 minutes and 9 seconds to complete, which is an extremely long execution time for a query that only needs to return 101 rows. This poor performance can be attributed to several factors:

1. Inefficient index usage: The optimizer chose the index on created_at, which resulted in scanning 25,147,450 rows.
2. Large intermediate result set: After applying the date range filter, 12,082,311 rows still needed to be processed.
3. Late filtering: The most selective predicates (mode, user_id, and label_id) were applied after accessing the table, resulting in 16,604 rows.
4. Sorting overhead: The final sort operation on 16,604 rows added additional processing time.

Here is the query pattern

> ```SQL

Reviewer 2: ````suggestion

```sql
````

SELECT `orders`.*
FROM `orders`
WHERE `orders`.`mode` = 'production'
AND `orders`.`user_id` = 11111
AND (orders.label_id IS NOT NULL)
AND (orders.created_at >= '2024-04-07 18:07:52')
AND (orders.created_at <= '2024-05-11 18:07:52')
AND (id >= 1000000000)
AND (id <= 2000000000)
AND (orders.id < 1500000000) 
ORDER BY orders.id DESC LIMIT 101;
```

```SQL
PRIMARY KEY (`id`),
UNIQUE KEY `index_orders_on_adjustment_id` (`adjustment_id`),
KEY `index_orders_on_user_id` (`user_id`),
KEY `index_orders_on_label_id` (`label_id`),
KEY `index_orders_on_created_at` (`created_at`)
```

> Original plan
> +--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
> | id                             | estRows   | actRows | task      | access object                                                                  | execution info                                      | operator info                                                                                                        | memory   | disk |
> +--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
> | TopN_10                        | 101.00    | 101     | root      |                                                                                | time:11m9.8s, loops:2                               | orders.id:desc, offset:0, count:101                                                                                  | 271 KB   | N/A  |
> | └─IndexLookUp_39               | 173.83    | 16604   | root      |                                                                                | time:11m9.8s, loops:19, index_task: {total_time:...}|                                                                                                                      | 20.4 MB  | N/A  |
> |   ├─Selection_37(Build)        | 8296.70   | 12082311| cop[tikv] |                                                                                | time:26.4ms, loops:11834, cop_task: {num: 294, m...}| ge(orders.id, 1000000000), le(orders.id, 2000000000), lt(orders.id, 1500000000)                                      | N/A      | N/A  |
> |   │ └─IndexRangeScan_35        | 6934161.90| 25147450| cop[tikv] | table:orders, index:index_orders_on_created_at(created_at)                     | tikv_task:{proc max:2.15s, min:0s, avg: 58.9ms, ...}| range:[2024-04-07 18:07:52,2024-05-11 18:07:52), keep order:false                                                    | N/A      | N/A  |
> |   └─Selection_38(Probe)        | 173.83    | 16604   | cop[tikv] |                                                                                | time:54m46.2s, loops:651, cop_task: {num: 1076, ...}| eq(orders.mode, "production"), eq(orders.user_id, 11111), not(isnull(orders.label_id))                               | N/A      | N/A  |
> |     └─TableRowIDScan_36        | 8296.70   | 12082311| cop[tikv] | table:orders                                                                   | tikv_task:{proc max:44.8s, min:0s, avg: 3.33s, p...}| keep order:false                                                                                                     | N/A      | N/A  |
> +--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+

Reviewer 2: ````suggestion
Original plan

```
+--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
| id                             | estRows   | actRows | task      | access object                                                                  | execution info                                      | operator info                                                                                                        | memory   | disk |
+--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
| TopN_10                        | 101.00    | 101     | root      |                                                                                | time:11m9.8s, loops:2                               | orders.id:desc, offset:0, count:101                                                                                  | 271 KB   | N/A  |
| └─IndexLookUp_39               | 173.83    | 16604   | root      |                                                                                | time:11m9.8s, loops:19, index_task: {total_time:...}|                                                                                                                      | 20.4 MB  | N/A  |
|   ├─Selection_37(Build)        | 8296.70   | 12082311| cop[tikv] |                                                                                | time:26.4ms, loops:11834, cop_task: {num: 294, m...}| ge(orders.id, 1000000000), le(orders.id, 2000000000), lt(orders.id, 1500000000)                                      | N/A      | N/A  |
|   │ └─IndexRangeScan_35        | 6934161.90| 25147450| cop[tikv] | table:orders, index:index_orders_on_created_at(created_at)                     | tikv_task:{proc max:2.15s, min:0s, avg: 58.9ms, ...}| range:[2024-04-07 18:07:52,2024-05-11 18:07:52), keep order:false                                                    | N/A      | N/A  |
|   └─Selection_38(Probe)        | 173.83    | 16604   | cop[tikv] |                                                                                | time:54m46.2s, loops:651, cop_task: {num: 1076, ...}| eq(orders.mode, "production"), eq(orders.user_id, 11111), not(isnull(orders.label_id))                               | N/A      | N/A  |
|     └─TableRowIDScan_36        | 8296.70   | 12082311| cop[tikv] | table:orders                                                                   | tikv_task:{proc max:44.8s, min:0s, avg: 3.33s, p...}| keep order:false                                                                                                     | N/A      | N/A  |
+--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
```
````

1. Efficient index usage: The new index allows for an index range scan on user_id, mode, and id, which are the most selective predicates. This drastically reduces the number of rows scanned from millions to just 224.
2. Index-only sort: The 'keep order:true' in the plan indicates that the index structure is used for sorting, avoiding a separate sort operation.
3. Early filtering: The most selective predicates are applied first, quickly reducing the result set to 224 rows before further filtering.
4. Limit push-down: The LIMIT clause is pushed down to the index scan, allowing early termination of the scan once 101 rows are found.

This case demonstrates the profound impact that a well-designed index can have on query performance. By aligning the index structure with the query's predicates, sort order, and required columns, we achieved a performance improvement of over five orders of magnitude.

> Plan with new index (user_id, mode, id, created_at, label_id) 
> +--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
> | id                             | estRows   | actRows | task      | access object                                                                  | execution info                                      | operator info                                                                                                        | memory   | disk |
> +--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
> | IndexLookUp_32                 | 101.00    | 101     | root      |                                                                                | time:5.3ms, loops:2, RU:3.435006, index_task: {t...}| limit embedded(offset:0, count:101)                                                                                  | 128.5 KB | N/A  |
> | ├─Limit_31(Build)              | 101.00    | 101     | cop[tikv] |                                                                                | time:1.35ms, loops:1, cop_task: {num: 1, max: 1....}| offset:0, count:101                                                                                                  | N/A      | N/A  |
> | │ └─Selection_30               | 535.77    | 224     | cop[tikv] |                                                                                | tikv_task:{time:0s, loops:3}                        | ge(orders.created_at, 2024-04-07 18:07:52), le(orders.created_at, 2024-05-11 18:07:52), not(isnull(orders.label_id)) | N/A      | N/A  |
> | │   └─IndexRangeScan_28        | 503893.42 | 224     | cop[tikv] | table:orders, index:index_orders_new(user_id, mode, id, created_at, label_id)  | tikv_task:{time:0s, loops:3}                        | range:[11111 "production" 1000000000,11111 "production" 1500000000), keep order:true, desc                           | N/A      | N/A  |
> | └─TableRowIDScan_29(Probe)     | 101.00    | 101     | cop[tikv] | table:orders                                                                   | time:2.9ms, loops:2, cop_task: {num: 3, max: 2.7...}| keep order:false                                                                                                     | N/A      | N/A  |
> +--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+

Reviewer 2: ````suggestion
Plan with new index (user_id, mode, id, created_at, label_id) 

```
+--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
| id                             | estRows   | actRows | task      | access object                                                                  | execution info                                      | operator info                                                                                                        | memory   | disk |
+--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
| IndexLookUp_32                 | 101.00    | 101     | root      |                                                                                | time:5.3ms, loops:2, RU:3.435006, index_task: {t...}| limit embedded(offset:0, count:101)                                                                                  | 128.5 KB | N/A  |
| ├─Limit_31(Build)              | 101.00    | 101     | cop[tikv] |                                                                                | time:1.35ms, loops:1, cop_task: {num: 1, max: 1....}| offset:0, count:101                                                                                                  | N/A      | N/A  |
| │ └─Selection_30               | 535.77    | 224     | cop[tikv] |                                                                                | tikv_task:{time:0s, loops:3}                        | ge(orders.created_at, 2024-04-07 18:07:52), le(orders.created_at, 2024-05-11 18:07:52), not(isnull(orders.label_id)) | N/A      | N/A  |
| │   └─IndexRangeScan_28        | 503893.42 | 224     | cop[tikv] | table:orders, index:index_orders_new(user_id, mode, id, created_at, label_id)  | tikv_task:{time:0s, loops:3}                        | range:[11111 "production" 1000000000,11111 "production" 1500000000), keep order:true, desc                           | N/A      | N/A  |
| └─TableRowIDScan_29(Probe)     | 101.00    | 101     | cop[tikv] | table:orders                                                                   | time:2.9ms, loops:2, cop_task: {num: 3, max: 2.7...}| keep order:false                                                                                                     | N/A      | N/A  |
+--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
```
````

|   └─StreamAgg_9          | 1.00        | cop[tikv]    |                   | funcs:count(1)->Column#13                                                                          |
|     └─Selection_19       | 250.00      | cop[tikv]    |                   | ge(trips.start_date, 2017-07-01 00:00:00.000000), le(trips.start_date, 2017-07-01 23:59:59.000000) |
|       └─TableFullScan_18 | 10000.00    | cop[tikv]    | table:trips       | keep order:false, stats:pseudo                                                                     |
+--------------------------+-------------+--------------+-------------------+----------------------------------------------------------------------------------------------------+
5 rows in set (0.00 sec)
```

Additional Information in [EXPLAIN ANALYZE](sql-statements/sql-statement-explain-analyze.md) Output
Different from `EXPLAIN`, `EXPLAIN ANALYZE` executes the corresponding SQL statement, records its runtime information, and returns the information together with the execution plan. There runtime information is crucial for debugging query execution.

Description

> - actRows: Number of rows output by the operator.

Reviewer 2: ```suggestion

- actRows: Number of rows output by the operator.
```

- execution info: Detailed execution information of the operator. time represents the total wall time from entering the operator to leaving the operator, including the total execution time of all sub-operators. If the operator is called many times by the parent operator then the time refers to the accumulated time. loops is the number of times the current operator is called by the parent operator.
- memory: Memory used by the operator.
- disk: Disk space used by the operator.

Note: Some attributes and explain table columns are omitted for improved formatting

> ```sql

Reviewer 2: ````suggestion

```sql
````

```SQL
EXPLAIN SELECT COUNT(*) FROM trips WHERE start_date BETWEEN '2017-07-01 00:00:00' AND '2017-07-01 23:59:59';

> 
> +--------------------------+-------------+--------------+-------------------+----------------------------------------------------------------------------------------------------+

Reviewer 2: ````suggestion
```

```
+--------------------------+-------------+--------------+-------------------+----------------------------------------------------------------------------------------------------+
````

|       └─TableFullScan_18 | 10000.00    | cop[tikv]    | table:trips       | keep order:false, stats:pseudo                                                                     |
+--------------------------+-------------+--------------+-------------------+----------------------------------------------------------------------------------------------------+
5 rows in set (0.00 sec)
```

Let's apply the "first child first – recursive descent" rule to the second plan. In the below example begin from the top to bottom, by looking at the `IndexRangeScan_47` (the first child of the tree). For the table `stars`, the optimizer ony need to select the column `name` and `id`, the two columns can be met by the index `name(name)`. So for the table `star`, the root reader is `IndexReader_48`, rather than a `TableReader`.
The join method between `stars` and `planets` is a hash join, which is marked as `HashJoin_44`. The data access method on `planets` is a `TableFullScan_45`. After the join, the `TopN_26` and `TOPN_19` is to implemented the two order by and limit corespondingly. The final operator `Projection_16` is to implemented the column projection for `t5.name`.

> ```SQL
> tidb> EXPLAIN SELECT t5.name FROM
>     -> (SELECT p.name, p.gravity, p.distance_from_sun FROM universe.planets p JOIN universe.stars s
>     ->  ON s.id = p.sun_id AND s.name = 'Sun'
>     ->  ORDER BY p.distance_from_sun ASC LIMIT 5) t5
>     -> ORDER BY t5.gravity DESC LIMIT 3;
> +-----------------------------------+----------+-----------+---------------------------+

Reviewer 2: ````suggestion
```sql
EXPLAIN SELECT t5.name FROM
(SELECT p.name, p.gravity, p.distance_from_sun FROM universe.planets p JOIN universe.stars s
ON s.id = p.sun_id AND s.name = 'Sun'
ORDER BY p.distance_from_sun ASC LIMIT 5) t5
ORDER BY t5.gravity DESC LIMIT 3;
```

```
+-----------------------------------+----------+-----------+---------------------------+
````

In TiDB, leveraging indexes effectively is crucial for performance tuning, as it can significantly reduce the number of KV [RPC](/glossary.md#remote-procedure-call-rpc) requests. By minimizing these requests, you can greatly improve query performance and overall system efficiency. Here are some key strategies:

- Avoiding full table scans
- Avoiding sorting
- Skipping row lookups when possible

This section showcases three practical examples that illustrate effective indexing strategies in TiDB, focusing on the use of composite and covered indexes for both filtering and sorting operations.

### SQL Tuning with a Covered Index

A covered index is designed to include all columns referenced in the filter and select clauses. The query below requires an index lookup of 2597411 rows, taking 46.4 seconds to execute.
TiDB needs to dispatch 67 cop tasks for the index range scan on logs_idx, identified as `IndexRangeScan_11`, and 301 cop tasks for table access via `TableRowIDScan_12`.
By utilizing a covered index, the index lookup can be avoided, leading to improved performance.

> ```SQL

Reviewer 2: ````suggestion
```sql
````

SELECT
  SUM(`logs`.`amount`)
FROM
  `logs`
WHERE
  `logs`.`user_id` = 1111
  AND `logs`.`snapshot_id` IS NULL
  AND `logs`.`status` IN ('complete', 'failure')
  AND `logs`.`source_type` != 'online'
  AND (
    `logs`.`source_type` IN ('user', 'payment')
    OR `logs`.`source_type` IN (
      'bank_account',
    )
    AND `logs`.`target_type` IN ('bank_account')
  );
```

> ```SQL

Reviewer 2: ````suggestion
```
````

---
title: SQL Tuning Best Practice
summary: Learn how to do SQL tuning in TiDB
---

# Introduction to SQL Tuning

SQL tuning is a critical aspect of optimizing database system performance. It involves a systematic approach to improve the efficiency of SQL queries. The process typically consists of three key steps:

1. Identify high-impact SQL statements:
   - Review SQL execution history to find statements that consume a large portion of system resources or contribute significantly to the application workload.
   - Use monitoring tools and performance metrics to pinpoint these resource-intensive queries.

2. Analyze execution plans:
   - Examine the execution plans generated by the query optimizer for the identified statements.
   - Verify that these plans are reasonably efficient and utilize appropriate indexes and join methods.

3. Implement optimizations:
   - Apply corrective actions to improve poorly performing SQL statements.
   - This may include rewriting queries, adding or modifying indexes, updating statistics, or adjusting database parameters.

These steps are iteratively repeated until:

> - The system performance meets the desired targets

Reviewer 2: ```suggestion

- The system performance meets the desired targets
```

- No further improvements can be made to the remaining statements.

It's important to note that SQL tuning is an ongoing process. As your data volumes grow and query patterns evolve, you should:

> - Regularly monitor query performance

Reviewer 2: ```suggestion

- Regularly monitor query performance
```

```
sql-tuning-best-practice.md: 27: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "- Regularly monitor query perf..."]
```
```

2. Enhancing data access methods:
   - Use covering indexes to satisfy queries directly from the index, avoiding table access
   - Implement partitioning strategies to limit data scans to relevant partitions

Examples:

- Creating an index for frequently queried columns can significantly reduce resource usage, especially for queries that access a small percentage of table data.
- Utilizing index-only scans for queries that return a limited number of sorted results can avoid full table scans and sorting operations.

## Balance Workload Distribution

In a distributed architecture like TiDB, it's essential to maintain a balanced workload across multiple TiKV nodes. This is crucial for achieving optimal performance. For guidance on identifying and resolving read and write hotspots, please refer to the [document](https://docs.pingcap.com/tidb/stable/troubleshoot-hot-spot-issues#optimization-of-small-table-hotspots). By implementing these strategies, you can ensure that your TiDB cluster efficiently utilizes all available resources and avoids bottlenecks caused by uneven workload distribution or serialization on individual TiKV nodes.

# Identifying High-Load SQL

> The most efficient way to identify Resource-Intensive SQL is using TiDB Dashboard, There are other tools like views and logs available as well.

Reviewer 2: ```suggestion

The most efficient way to identify Resource-Intensive SQL is using TiDB Dashboard, There are other tools like views and logs available as well.
```

```
sql-tuning-best-practice.md: 64: MD022/blanks-around-headings/blanks-around-headers Headings should be surrounded by blank lines [Expected: 1; Actual: 0; Below] [Context: "# Identifying High-Load SQL"]
```

## Monitoring SQL Statements by Using TiDB Dashboard
### SQL Statements Panel 
In [TiDB Dashboard](dashboard/dashboard-overview.md), navigate to SQL Statements panel, which helps us identify the following:

1. The SQL statement with the highest total latency is the one that takes the longest time to execute out of multiple executions of the same SQL statement. 
2. It can also display the number of times each SQL statement has been executed cumulatively, allowing us to find the SQL statement with the highest execution frequency.
3. Clicking on each SQL statement allows us to delve deeper into the `EXPLAIN ANALYZE` results.

SQL statements are normalized as templates, where literals and bind variables are replaced by '?'. This normalization and sorting process allows you to quickly pinpoint the most resource-intensive queries that may require optimization.
![sql-statements-default](/media/sql-tuning/sql-statements-default.png)

### Slow Queries Panel Default Display

> In the TiDB Dashboard, we can find the Slow Query panel, which displays all SQL statements whose execution time exceeds the time threshold set by the system variable tidb_slow_log_threshold (default 300 milliseconds). 

Reviewer 2: ```suggestion

In the TiDB Dashboard, we can find the Slow Query panel, which displays all SQL statements whose execution time exceeds the time threshold set by the system variable tidb_slow_log_threshold (default 300 milliseconds). 
```

```
sql-tuning-best-practice.md: 79: MD022/blanks-around-headings/blanks-around-headers Headings should be surrounded by blank lines [Expected: 1; Actual: 0; Below] [Context: "### Slow Queries Panel Default Display"]
```

On Slow Queries panel, we can find:

1. The slowest SQL queries.
2. The SQL query that reads the most data from TiKV.
3. The `EXPLAIN ANALYZE` output from drilling down the SQL statement by clicking it.
4. Please note that on the Slow Queries panel, we cannot get the frequency of the SQL statement execution. Once the execution elapsed time exceeds tidb_slow_log_threshold for single instance, the query is then listed on the Slow Queries panel.
![slow-query-default](/media/sql-tuning/slow-query-default.png)

## Other Tools for Identifying Top SQL

In addition to TiDB Dashboard, there are several other tools available to identify resource-intensive SQL queries:

Each tool offers unique insights and can be valuable for different analysis scenarios. Using a combination of these tools allows for comprehensive SQL performance monitoring and optimization.

> - [slow query log](https://docs.pingcap.com/tidb/stable/identify-slow-queries)

Reviewer 2: ```suggestion

- [slow query log](https://docs.pingcap.com/tidb/stable/identify-slow-queries)
```

```
sql-tuning-best-practice.md: 95: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "- [slow query log](https://doc/..."]
```

- [statements_summary view](https://docs.pingcap.com/tidb/stable/statement-summary-tables#statements_summary)
- [top sql feature](https://docs.pingcap.com/tidb/stable/top-sql)
- [expensive queries in tidb log](https://docs.pingcap.com/tidb/stable/identify-expensive-queries)
- [cluster_processlist view](https://docs.pingcap.com/tidb/stable/information-schema-processlist#cluster_processlist)

> 

Reviewer 2: ```suggestion
```

```
sql-tuning-best-practice.md: 101: MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
```

|         │ │ └─IndexRangeScan_38         |.| 1       | cop[tikv] | table:c, index:name(name) | tikv_task:{time:0s, loops:1}, scan_detail: {...}               |.| N/A       | N/A     |
|         │ └─TableReader_37(Probe)       |.| 10      | root      |                           | time:543.7µs, loops:2, cop_task: {...}                         |.| 577 Bytes | N/A     |
|         │   └─TableFullScan_36          |.| 10      | cop[tikv] | table:p                   | tikv_task:{time:0s, loops:1}, scan_detail: {...}               |.| N/A       | N/A     |
|         └─TableReader_22(Probe)         |.| 28      | root      |                           | time:671.7µs, loops:2, cop_task: {...}                         |.| 876 Bytes | N/A     |
|           └─TableFullScan_21            |.| 28      | cop[tikv] | table:m                   | tikv_task:{time:0s, loops:1}, scan_detail: {...}               |.| N/A       | N/A     |
+-----------------------------------------+.+---------+-----------+---------------------------+----------------------------------------------------------------+.+-----------+---------+
```

> 

Reviewer 2: ```suggestion
```

There are three important details to add to this:

1. Parent-Child Interaction: Although a parent operator calls its child operators in sequence, it may cycle through them multiple times. For example, in an index lookup or nested loop join, the parent fetches a batch of rows from the first child, then (zero or more) rows from the second child, repeating this process until it consumes the entire result set from the first child.

2. Blocking vs. Non-blocking Operators: Operators can be either blocking or non-blocking. Blocking operators, such as `TopN` and `HashAgg`, must create their entire result set before passing anything to their parent. Non-blocking operators, like `IndexLookup` and `IndexJoin`, create and pass their row source piece by piece on demand.

3. Concurrent vs. Serial Execution: Child operators can be executed concurrently or serially. For instance, the child operators of an `IndexLookup` operator are executed serially, while those of a `HashJoin` operator can be executed concurrently.

> 

Reviewer 2: ```suggestion
```

Let's apply the "first child first – recursive descent" rule to the first plan. When reading an execution plan, you should start from the top and work your down bottom. In the below example begin by looking at the `TableFullScan_18` (or the first child of the tree). In this case the access operator for table trips are implemented using full table scan. The rows produced by the tables scans will be consumed by the `Selection_19` operator. The `Selection_19` operator is to filter the data by `ge(trips.start_date, 2017-07-01 00:00:00.000000), le(trips.start_date, 2017-07-01 23:59:59.000000)`. Next the group-by operator `StreamAgg_9` is to implemented the aggregation `count(*)`. Be noted that the 3 operators `TableFullScan_18`, `Selection_19`, `StreamAgg_9` are pushdown to TiKV, which is marked as `cop[tikv]`, so that early filter and aggregation can be done in TiKV, to minize the data transfer between TiKV and TiDB. Finally the `TableReader_21` is to read the data from the `StreamAgg_9` operator, then finally the `StreamAgg_20` is to implemented the aggregation `count(*)`.

> 

Reviewer 2: ```suggestion
```

|       └─TableFullScan_18 | 10000.00    | cop[tikv]    | table:trips       | keep order:false, stats:pseudo                                                                     |
+--------------------------+-------------+--------------+-------------------+----------------------------------------------------------------------------------------------------+
5 rows in set (0.00 sec)
```

Let's apply the "first child first – recursive descent" rule to the second plan. In the below example begin from the top to bottom, by looking at the `IndexRangeScan_47` (the first child of the tree). For the table `stars`, the optimizer ony need to select the column `name` and `id`, the two columns can be met by the index `name(name)`. So for the table `star`, the root reader is `IndexReader_48`, rather than a `TableReader`.
The join method between `stars` and `planets` is a hash join, which is marked as `HashJoin_44`. The data access method on `planets` is a `TableFullScan_45`. After the join, the `TopN_26` and `TOPN_19` is to implemented the two order by and limit corespondingly. The final operator `Projection_16` is to implemented the column projection for `t5.name`.

> 

Reviewer 2: ```suggestion
```

   - Ensure sort and limit pushdown to TiKV
   - Maintain sorted order

3. Additional filtering columns:
   - Non-equal predicates (!=, <>, IS NOT NULL, etc)
   - Time range conditions on datetime columns
   - Helps reduce row lookups

4. columns in index postfix for select list or aggregate function:
   - Leverage IndexReader
   - Avoid IndexLookup operations

### The Cost of Indexing

While indexes can significantly improve query performance, they also come with costs that should be carefully considered:

1. Performance Impact on Writes:
   - Each additional index slows down write operations (INSERT, UPDATE, DELETE)
   - When data is modified, all affected indexes must be updated
   - This overhead increases with each additional index

2. Resource Consumption:
   - Indexes require additional disk space
   - More memory is needed to cache frequently accessed indexes
   - Backup and recovery operations take longer

3. Write Hotspot Risk:
   - Secondary indexes can create write hotspots, for example, a datetime monotonically increasing index will cause the hotspots on the write of the table
   - Performance can degrade significantly if hotspots occur

Best Practice:

> - Only create indexes that provide clear performance benefits

Reviewer 2: ```suggestion

- Only create indexes that provide clear performance benefits
```

```
sql-tuning-best-practice.md: 747: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "- Only create indexes that pro..."]
```

+--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
```

### Composite Index Strategy Guidelines

When creating a composite index, it's essential to follow a specific order for the columns to ensure optimal performance. This order is crucial because it directly impacts how efficiently the index can filter and sort data.

Here is the recommended order for the columns in the index:

1. Equal predicates for index prefix (columns accessed directly):
   - Equal conditions 
   - IS NULL conditions

2. Columns after index prefix for sorting:
   - Leverage index to preprocess sorting
   - Ensure sort and limit pushdown to TiKV
   - Maintain sorted order

3. Additional filtering columns:

>    - Non-equal predicates (!=, <>, IS NOT NULL, etc)

Reviewer 2: ```suggestion
   - Non-equal predicates (`!=`, `<>`, `IS NOT NULL`, etc)
```

I think this is better for formatting and might resolve the CI check

Examples:

- Creating an index for frequently queried columns can significantly reduce resource usage, especially for queries that access a small percentage of table data.
- Utilizing index-only scans for queries that return a limited number of sorted results can avoid full table scans and sorting operations.

## Balance Workload Distribution

In a distributed architecture like TiDB, it's essential to maintain a balanced workload across multiple TiKV nodes. This is crucial for achieving optimal performance. For guidance on identifying and resolving read and write hotspots, please refer to the [document](https://docs.pingcap.com/tidb/stable/troubleshoot-hot-spot-issues#optimization-of-small-table-hotspots). By implementing these strategies, you can ensure that your TiDB cluster efficiently utilizes all available resources and avoids bottlenecks caused by uneven workload distribution or serialization on individual TiKV nodes.

# Identifying High-Load SQL

The most efficient way to identify Resource-Intensive SQL is using TiDB Dashboard, There are other tools like views and logs available as well.

## Monitoring SQL Statements by Using TiDB Dashboard

### SQL Statements Panel 

> In [TiDB Dashboard](dashboard/dashboard-overview.md), navigate to SQL Statements panel, which helps us identify the following:

Reviewer 2: ```suggestion

In [TiDB Dashboard](dashboard/dashboard-overview.md), navigate to SQL Statements panel, which helps us identify the following:
```

```
sql-tuning-best-practice.md: 72: MD022/blanks-around-headings/blanks-around-headers Headings should be surrounded by blank lines [Expected: 1; Actual: 0; Below] [Context: "### SQL Statements Panel"]
```

1. The SQL statement with the highest total latency is the one that takes the longest time to execute out of multiple executions of the same SQL statement. 
2. It can also display the number of times each SQL statement has been executed cumulatively, allowing us to find the SQL statement with the highest execution frequency.
3. Clicking on each SQL statement allows us to delve deeper into the `EXPLAIN ANALYZE` results.

SQL statements are normalized as templates, where literals and bind variables are replaced by '?'. This normalization and sorting process allows you to quickly pinpoint the most resource-intensive queries that may require optimization.
![sql-statements-default](/media/sql-tuning/sql-statements-default.png)

> 

Reviewer 2: ```suggestion
```

```
sql-tuning-best-practice.md: 82: MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
```

- [cluster_processlist view](https://docs.pingcap.com/tidb/stable/information-schema-processlist#cluster_processlist)

## Gathering Data on the SQL Identified

For the top SQL statements identified, you can use [PLAN REPLAYER](/sql-plan-replayer.md) to capture and save the on-site information of a TiDB cluster. This tool allows you to recreate the execution environment for further analysis. The syntax for exporting SQL information is as follows:

```sql
PLAN REPLAYER DUMP EXPLAIN [ANALYZE] [WITH STATS AS OF TIMESTAMP expression] sql-statement;
```

Use `EXPLAIN ANALYZE` whenever possible, as it captures actual execution information in addition to the execution plan. This provides more accurate insights into query performance.

# SQL Tuning Guide

This guide focuses on providing actionable advice for beginners looking to optimize their SQL queries in TiDB. By following these best practices, you can ensure better query performance and SQL Tuning. We'll cover below topic:

- Query Processing Workflow
- Optimizer Fundamentals
- Statistics Management
- How TiDB build A Execution Plan
- Understand Execution Plan
- Index Strategy in TiDB

>   - SQL Tuning with a Covered Index

Reviewer 2: ```
sql-tuning-best-practice.md: 127: MD007/ul-indent Unordered list indentation [Expected: 4; Actual: 2]
```

2. Pre-Process is primarily for Point Get. If it is a Point Get, the following-up optimization processes can be skipped, the next step jumps to the SQL Executor.
3. After confirming that it is not a Point Get, the AST goes to the Logical Transformation, which rewrites the SQL logically based on certain rules.
4. The AST that has gone through Logical Transformation will undergo Cost-Based Optimization.
5. During Cost-Based Optimization, the optimizer considers statistics to determine how to select specific operators and finally generates a physically executable physical execution plan.
6. The generated physical execution plan is sent to the SQL Executor of the TiDB database for execution.
7. Unlike traditional databases, TiDB databases need to push down the execution plan to different TiKV for reading and writing data.

![workflow](/media/sql-tuning/workflow.png)

## Optimizer Fundamentals

TiDB uses a cost-based optimizer (CBO) to determine the most efficient execution plan for a SQL statement. This optimizer evaluates different execution strategies and chooses the one with the lowest estimated cost. The cost is influenced by factors such as:

- SQL
- Schema Design
- Statistics

>   - Table

Reviewer 2: ```
sql-tuning-best-practice.md: 156: MD007/ul-indent Unordered list indentation [Expected: 4; Actual: 2]
```

  - Index
  - Column

Based on the input, The cost model will produce the execution plan, which includes the details how the system execute the sql, including 

> - Access Method

Reviewer 2: ```suggestion

- Access Method
```

```
sql-tuning-best-practice.md: 161: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "- Access Method"]
```

- Join Method
- Join Order

The optimizer is as good as the information it receives. Therefore, ensuring up-to-date statistics and well-designed indexes is critical.

## Statistics Management

Statistics are essential to the TiDB optimizer. TiDB uses statistics as input to the optimizer to estimate the number of rows processed in each plan step for a SQL statement.

statistics is generally divided into two levels: table level and column level. 

> - For table-level statistics, it includes the total number of rows in the table and the number of rows that have been modified since the last collection of statistics. 

Reviewer 2: ```suggestion

- For table-level statistics, it includes the total number of rows in the table and the number of rows that have been modified since the last collection of statistics. 
```

```
sql-tuning-best-practice.md: 172: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "- For table-level statistics, ..."]
```

1. [`SHOW STATS_META`](sql-statements/sql-statement-show-stats-meta.md): This command provides metadata about table statistics.
2. [`SHOW STATS_HEALTHY`](sql-statements/sql-statement-show-stats-healthy.md): This command shows the health status of table statistics.

For example, you can use:

```sql
SHOW STATS_META WHERE table_name='T2'\G;
```

```
*************************** 1. row ***************************
 Db_name: test
 Table_name: T2
Partition_name:
 Update_time: 2023-05-11 02:16:50
 Modify_count: 20000
 Row_count: 20000
1 row in set (0.03 sec)
```

```sql
SHOW STATS_HEALTHY WHERE table_name='T2'\G;
```

```
*************************** 1. row ***************************
    Db_name: test
  Table_name: T2
Partition_name:
    Healthy: 0
1 row in set (0.00 sec)
```

In TiDB database, there are two ways to collect statistics: automatic collection and manual collection. In most case, the auto collection job works fine. Automatic collection is actually triggered when certain conditions are met for a table, and TiDB will automatically collect statistics. We commonly use three triggering conditions, which are: ratio, start_time and end_time.

> - tidb_auto_analyze_ratio: The healthiness trigger

Reviewer 2: ```suggestion

- tidb_auto_analyze_ratio: The healthiness trigger
```

```
sql-tuning-best-practice.md: 211: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "- tidb_auto_analyze_ratio: The..."]
```

```sql
SHOW VARIABLES LIKE 'tidb\_auto\_analyze%';
```

```
+-----------------------------------------+-------------+
| Variable_name                           | Value       |
+-----------------------------------------+-------------+
| tidb_auto_analyze_ratio                 | 0.5         |
| tidb_auto_analyze_start_time            | 00:00 +0000 |
| tidb_auto_analyze_end_time              | 23:59 +0000 |
+-----------------------------------------+-------------+
```

In cases where automatic collection doesn't meet your needs, you can manually collect statistics using the `ANALYZE TABLE table_name` statement. This allows you to:

1. Adjust the sample rate for more accurate or faster analysis
2. Increase the number of top-N values collected
3. Gather statistics for specific columns only

It's important to note that after manual collection, subsequent automatic gathering jobs will inherit the new parameters. This means that any customizations you've made during manual collection will be carried forward in future automatic analyses.

Another common scenario is locking table statistics. This is useful when:

> 1. The statistics on the table are already representative of the data.

Reviewer 2: ```suggestion

1. The statistics on the table are already representative of the data.
```

```
sql-tuning-best-practice.md: 237: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "1. The statistics on the table..."]
```

2. The table is very large and statistics collection is time-consuming.
3. You want to maintain statistics only during specific time windows.

To lock the statistics for a table, you can use the following statement [`LOCK STATS table_name`](/sql-statements/sql-statement-lock-stats.md).

for more detail about statistics, please refer to [statistics](https://docs.pingcap.com/tidb/stable/statistics).

## How TiDB build A Execution Plan

> An SQL statement undergoes optimization primarily in the optimizer through three stages:

Reviewer 2: ```suggestion

An SQL statement undergoes optimization primarily in the optimizer through three stages:
```

```
sql-tuning-best-practice.md: 245: MD022/blanks-around-headings/blanks-around-headers Headings should be surrounded by blank lines [Expected: 1; Actual: 0; Below] [Context: "## How TiDB build A Execution Plan"]
```

Point_Get or Batch_Point_Get is to get 1 or 0 or many row only by using the TiKV key, the explicit or implicit (`_tidb_rowid`) primary key. For example, when id column is the primary key of a clustered index table, Point_Get is used to get the particular row. If a plan is identified as Point_Get, optimizer will skip the logical transformation and cost-based optimization.

```sql
SELECT id, name FROM emp WHERE id = 901; 
```

### Logical Transformation

The purpose of logical Transformation is to optimize the execution of statements based on the characteristics of SELECT list, WHERE predicates, and other predicates in SQL queries. It generates a logical execution plan to annotate and rewrite the query. This logical plan is then passed to the Cost-Based Optimization. The optimization rules include column pruning，partition pruning，eliminate Max/Min, eliminate outer join, join reorder, predicates push-down，subquery rewrite，derive TopN from window functions，and de-correlation of correlated Subquery. Since this step is fully automated by the query optimizer, it usually does not require manual adjustments.

> More Detail for Logical Transformation: https://docs.pingcap.com/tidb/stable/sql-logical-optimization.

Reviewer 2: ```suggestion
More Detail for Logical Transformation: [SQL Logical Optimization](/sql-logical-optimization.md).
```

```
sql-tuning-best-practice.md: 266: MD034/no-bare-urls Bare URL used [Context: "https://docs.pingcap.com/tidb/..."]
```

Furthermore, confirming the algorithm for certain SQL operations, such as aggregation, join, and sorting, is essential. For instance, the aggregation operator may utilize either `HASH_AGG` or `STREAM_AGG`, while the join operator can select from `HASH JOIN`, `MERGE JOIN`, or `INDEX JOIN`. Likewise, various options are available for the sorting operator.

![cost-based-optimization](/media/sql-tuning/cost-based-optimization.png)

## Understanding Execution Plans

The execution plan represents the steps TiDB will follow to execute a SQL query. In this section, we will learn how to display and read the execution plan.

### Generating and Displaying Execution Plans

Beside access the execution plan information through TiDB Dashboard, TiDB provides a `EXPLAIN` statement to display the execution plan for a SQL query. Here's an example of using `EXPLAIN`:

> - id: Operator name and the step unique identifier

Reviewer 2: ```suggestion

- id: Operator name and the step unique identifier
```

```
sql-tuning-best-practice.md: 285: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "- id: Operator name and the st..."]
```

There are three important details to add to this:

1. Parent-Child Interaction: Although a parent operator calls its child operators in sequence, it may cycle through them multiple times. For example, in an index lookup or nested loop join, the parent fetches a batch of rows from the first child, then (zero or more) rows from the second child, repeating this process until it consumes the entire result set from the first child.

2. Blocking vs. Non-blocking Operators: Operators can be either blocking or non-blocking. Blocking operators, such as `TopN` and `HashAgg`, must create their entire result set before passing anything to their parent. Non-blocking operators, like `IndexLookup` and `IndexJoin`, create and pass their row source piece by piece on demand.

3. Concurrent vs. Serial Execution: Child operators can be executed concurrently or serially. For instance, the child operators of an `IndexLookup` operator are executed serially, while those of a `HashJoin` operator can be executed concurrently.

> 

Reviewer 2: ```suggestion
```

```
sql-tuning-best-practice.md: 363: MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
```

+----------------------------------+---------+---------+-----------+----------------------------------------------+----------------------------------------------+----------------------------------------------------+
```

### SQL Tuning with Composite Indexes for Efficient Filtering and Sorting

The original query took 11 minutes and 9 seconds to complete, which is an extremely long execution time for a query that only needs to return 101 rows. This poor performance can be attributed to several factors:

1. Inefficient index usage: The optimizer chose the index on created_at, which resulted in scanning 25,147,450 rows.
2. Large intermediate result set: After applying the date range filter, 12,082,311 rows still needed to be processed.
3. Late filtering: The most selective predicates (mode, user_id, and label_id) were applied after accessing the table, resulting in 16,604 rows.
4. Sorting overhead: The final sort operation on 16,604 rows added additional processing time.

> 

Reviewer 2: ```suggestion
```

```
sql-tuning-best-practice.md: 638: MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
```

|   └─Selection_38(Probe)        | 173.83    | 16604   | cop[tikv] |                                                                                | time:54m46.2s, loops:651, cop_task: {num: 1076, ...}| eq(orders.mode, "production"), eq(orders.user_id, 11111), not(isnull(orders.label_id))                               | N/A      | N/A  |
|     └─TableRowIDScan_36        | 8296.70   | 12082311| cop[tikv] | table:orders                                                                   | tikv_task:{proc max:44.8s, min:0s, avg: 3.33s, p...}| keep order:false                                                                                                     | N/A      | N/A  |
+--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
```

> 

Reviewer 2: ```suggestion
```

```
sql-tuning-best-practice.md: 678: MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
```

| │   └─IndexRangeScan_28        | 503893.42 | 224     | cop[tikv] | table:orders, index:index_orders_new(user_id, mode, id, created_at, label_id)  | tikv_task:{time:0s, loops:3}                        | range:[11111 "production" 1000000000,11111 "production" 1500000000), keep order:true, desc                           | N/A      | N/A  |
| └─TableRowIDScan_29(Probe)     | 101.00    | 101     | cop[tikv] | table:orders                                                                   | time:2.9ms, loops:2, cop_task: {num: 3, max: 2.7...}| keep order:false                                                                                                     | N/A      | N/A  |
+--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
```

> 

Reviewer 2: ```suggestion
```

```
sql-tuning-best-practice.md: 704: MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
```

### Composite Index Strategy Guidelines

When creating a composite index, it's essential to follow a specific order for the columns to ensure optimal performance. This order is crucial because it directly impacts how efficiently the index can filter and sort data.
Here is the recommended order for the columns in the index:

1. Equal predicates for index prefix (columns accessed directly):
   - Equal conditions 
   - IS NULL conditions

2. Columns after index prefix for sorting:
   - Leverage index to preprocess sorting
   - Ensure sort and limit pushdown to TiKV
   - Maintain sorted order

3. Additional filtering columns:
   - Non-equal predicates (!=, <>, IS NOT NULL, etc)
   - Time range conditions on datetime columns
   - Helps reduce row lookups

4. columns in index postfix for select list or aggregate function:
   - Leverage IndexReader
   - Avoid IndexLookup operations

> 

Reviewer 2: ```suggestion
```

```
sql-tuning-best-practice.md: 728: MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
```

2. Minimize resource consumption for processing workloads

These goals can be achieved through various strategies:

## Optimize Query Execution

SQL tuning often involves finding more efficient ways to process the same workload without changing the query's functionality. This can be done by:

1. Improving execution plans:
   - Analyze and modify query structures to enable more efficient processing
   - Utilize appropriate indexes to reduce data access and processing time
   - Enable TiFlash for analytical queries on large volumes of data and Leverage MPP (Massively Parallel Processing) engine for complex aggregations and joins

2. Enhancing data access methods:
   - Use covering indexes to satisfy queries directly from the index, avoiding table access
   - Implement partitioning strategies to limit data scans to relevant partitions

Examples:

- Creating an index for frequently queried columns can significantly reduce resource usage, especially for queries that access a small percentage of table data.
- Utilizing index-only scans for queries that return a limited number of sorted results can avoid full table scans and sorting operations.

## Balance Workload Distribution

> In a distributed architecture like TiDB, it's essential to maintain a balanced workload across multiple TiKV nodes. This is crucial for achieving optimal performance. For guidance on identifying and resolving read and write hotspots, please refer to the [document](https://docs.pingcap.com/tidb/stable/troubleshoot-hot-spot-issues#optimization-of-small-table-hotspots). By implementing these strategies, you can ensure that your TiDB cluster efficiently utilizes all available resources and avoids bottlenecks caused by uneven workload distribution or serialization on individual TiKV nodes.

Reviewer 2: ```suggestion
In a distributed architecture like TiDB, it's essential to maintain a balanced workload across multiple TiKV nodes. This is crucial for achieving optimal performance. For guidance on identifying and resolving read and write hotspots, please refer to the [document](/troubleshoot-hot-spot-issues.md#optimization-of-small-table-hotspots). By implementing these strategies, you can ensure that your TiDB cluster efficiently utilizes all available resources and avoids bottlenecks caused by uneven workload distribution or serialization on individual TiKV nodes.
```

![sql-statements-default](/media/sql-tuning/sql-statements-default.png)

### Slow Queries Panel Default Display

In the TiDB Dashboard, we can find the Slow Query panel, which displays all SQL statements whose execution time exceeds the time threshold set by the system variable tidb_slow_log_threshold (default 300 milliseconds). 

On Slow Queries panel, we can find:

1. The slowest SQL queries.
2. The SQL query that reads the most data from TiKV.
3. The `EXPLAIN ANALYZE` output from drilling down the SQL statement by clicking it.
4. Please note that on the Slow Queries panel, we cannot get the frequency of the SQL statement execution. Once the execution elapsed time exceeds tidb_slow_log_threshold for single instance, the query is then listed on the Slow Queries panel.
![slow-query-default](/media/sql-tuning/slow-query-default.png)

## Other Tools for Identifying Top SQL

In addition to TiDB Dashboard, there are several other tools available to identify resource-intensive SQL queries:

Each tool offers unique insights and can be valuable for different analysis scenarios. Using a combination of these tools allows for comprehensive SQL performance monitoring and optimization.

> - [slow query log](https://docs.pingcap.com/tidb/stable/identify-slow-queries)
> - [statements_summary view](https://docs.pingcap.com/tidb/stable/statement-summary-tables#statements_summary)
> - [top sql feature](https://docs.pingcap.com/tidb/stable/top-sql)
> - [expensive queries in tidb log](https://docs.pingcap.com/tidb/stable/identify-expensive-queries)
> - [cluster_processlist view](https://docs.pingcap.com/tidb/stable/information-schema-processlist#cluster_processlist)

Reviewer 2: ```suggestion
- [slow query log](/identify-slow-queries.md)
- [statements_summary view](/statement-summary-tables.md#statements_summary)
- [Top SQL feature](/dashboard/top-sql.md)
- [expensive queries in TiDB log](/identify-expensive-queries.md)
- [cluster_processlist view](/information-schema/information-schema-processlist.md#cluster_processlist)
```

| tidb_auto_analyze_start_time            | 00:00 +0000 |
| tidb_auto_analyze_end_time              | 23:59 +0000 |
+-----------------------------------------+-------------+
```

In cases where automatic collection doesn't meet your needs, you can manually collect statistics using the `ANALYZE TABLE table_name` statement. This allows you to:

1. Adjust the sample rate for more accurate or faster analysis
2. Increase the number of top-N values collected
3. Gather statistics for specific columns only

It's important to note that after manual collection, subsequent automatic gathering jobs will inherit the new parameters. This means that any customizations you've made during manual collection will be carried forward in future automatic analyses.

Another common scenario is locking table statistics. This is useful when:

1. The statistics on the table are already representative of the data.
2. The table is very large and statistics collection is time-consuming.
3. You want to maintain statistics only during specific time windows.

To lock the statistics for a table, you can use the following statement [`LOCK STATS table_name`](/sql-statements/sql-statement-lock-stats.md).

> for more detail about statistics, please refer to [statistics](https://docs.pingcap.com/tidb/stable/statistics).

Reviewer 2: ```suggestion
for more detail about statistics, please refer to [statistics](/statistics.md).
```

| StreamAgg_20             | 1.00        | root         |                   | funcs:count(Column#13)->Column#11                                                                  |
| └─TableReader_21         | 1.00        | root         |                   | data:StreamAgg_9                                                                                   |
|   └─StreamAgg_9          | 1.00        | cop[tikv]    |                   | funcs:count(1)->Column#13                                                                          |
|     └─Selection_19       | 250.00      | cop[tikv]    |                   | ge(trips.start_date, 2017-07-01 00:00:00.000000), le(trips.start_date, 2017-07-01 23:59:59.000000) |
|       └─TableFullScan_18 | 10000.00    | cop[tikv]    | table:trips       | keep order:false, stats:pseudo                                                                     |
+--------------------------+-------------+--------------+-------------------+----------------------------------------------------------------------------------------------------+
5 rows in set (0.00 sec)
```

> Additional Information in [EXPLAIN ANALYZE](sql-statements/sql-statement-explain-analyze.md) Output
> Different from `EXPLAIN`, `EXPLAIN ANALYZE` executes the corresponding SQL statement, records its runtime information, and returns the information together with the execution plan. There runtime information is crucial for debugging query execution.

Reviewer 2: ```suggestion
Additional Information in [EXPLAIN ANALYZE](sql-statements/sql-statement-explain-analyze.md) Output. Different from `EXPLAIN`, `EXPLAIN ANALYZE` executes the corresponding SQL statement, records its runtime information, and returns the information together with the execution plan. There runtime information is crucial for debugging query execution.
```

```suggestion
Additional Information in [EXPLAIN ANALYZE](sql-statements/sql-statement-explain-analyze.md) Output

Different from `EXPLAIN`, `EXPLAIN ANALYZE` executes the corresponding SQL statement, records its runtime information, and returns the information together with the execution plan. There runtime information is crucial for debugging query execution.
```

```
MANUAL LINE BREAKS: L315
```

Description

- actRows: Number of rows output by the operator.
- execution info: Detailed execution information of the operator. time represents the total wall time from entering the operator to leaving the operator, including the total execution time of all sub-operators. If the operator is called many times by the parent operator then the time refers to the accumulated time. loops is the number of times the current operator is called by the parent operator.
- memory: Memory used by the operator.
- disk: Disk space used by the operator.

Note: Some attributes and explain table columns are omitted for improved formatting

```sql
EXPLAIN ANALYZE
SELECT SUM(pm.m_count)/COUNT(*) FROM
(SELECT COUNT(m.name) m_count
FROM universe.moons m
RIGHT JOIN
(SELECT p.id, p.name
FROM universe.planet_categories c
JOIN universe.planets p
ON c.id = p.category_id AND c.name = 'Jovian') pc
ON m.planet_id = pc.id
GROUP BY pc.name) pm;

> +-----------------------------------------+.+---------+-----------+---------------------------+----------------------------------------------------------------+.+-----------+---------+

Reviewer 2: ````suggestion
```

```
+-----------------------------------------+.+---------+-----------+---------------------------+----------------------------------------------------------------+.+-----------+---------+
````

|         │ │ └─IndexRangeScan_38         |.| 1       | cop[tikv] | table:c, index:name(name) | tikv_task:{time:0s, loops:1}, scan_detail: {...}               |.| N/A       | N/A     |
|         │ └─TableReader_37(Probe)       |.| 10      | root      |                           | time:543.7µs, loops:2, cop_task: {...}                         |.| 577 Bytes | N/A     |
|         │   └─TableFullScan_36          |.| 10      | cop[tikv] | table:p                   | tikv_task:{time:0s, loops:1}, scan_detail: {...}               |.| N/A       | N/A     |
|         └─TableReader_22(Probe)         |.| 28      | root      |                           | time:671.7µs, loops:2, cop_task: {...}                         |.| 876 Bytes | N/A     |
|           └─TableFullScan_21            |.| 28      | cop[tikv] | table:m                   | tikv_task:{time:0s, loops:1}, scan_detail: {...}               |.| N/A       | N/A     |
+-----------------------------------------+.+---------+-----------+---------------------------+----------------------------------------------------------------+.+-----------+---------+
```

### Reading Execution Plans: First Child First

> To understand why SQL queries run slowly, it's very important to know how to read Execution Plans. The main rule for reading an execution plan is "first child first – recursive descent".
> Each operator of the plan produces rows of data. When we talk about how a plan runs, we really mean the order in which each operator produces its rows. The "first child first" rule means that to produce its rows, each operator of the plan asks its child operators to produce their rows first. Then it combines these rows in some way. The order in which it asks its child parts is the same as the order they appear in the plan.

Reviewer 2: ```suggestion
To understand why SQL queries run slowly, it's very important to know how to read Execution Plans. The main rule for reading an execution plan is "first child first – recursive descent". Each operator of the plan produces rows of data. When we talk about how a plan runs, we really mean the order in which each operator produces its rows. The "first child first" rule means that to produce its rows, each operator of the plan asks its child operators to produce their rows first. Then it combines these rows in some way. The order in which it asks its child parts is the same as the order they appear in the plan.
```

```
MANUAL LINE BREAKS: L359
```

| StreamAgg_20             | 1.00        | root         |                   | funcs:count(Column#13)->Column#11                                                                  |
| └─TableReader_21         | 1.00        | root         |                   | data:StreamAgg_9                                                                                   |
|   └─StreamAgg_9          | 1.00        | cop[tikv]    |                   | funcs:count(1)->Column#13                                                                          |
|     └─Selection_19       | 250.00      | cop[tikv]    |                   | ge(trips.start_date, 2017-07-01 00:00:00.000000), le(trips.start_date, 2017-07-01 23:59:59.000000) |
|       └─TableFullScan_18 | 10000.00    | cop[tikv]    | table:trips       | keep order:false, stats:pseudo                                                                     |
+--------------------------+-------------+--------------+-------------------+----------------------------------------------------------------------------------------------------+
5 rows in set (0.00 sec)
```

> Let's apply the "first child first – recursive descent" rule to the second plan. In the below example begin from the top to bottom, by looking at the `IndexRangeScan_47` (the first child of the tree). For the table `stars`, the optimizer ony need to select the column `name` and `id`, the two columns can be met by the index `name(name)`. So for the table `star`, the root reader is `IndexReader_48`, rather than a `TableReader`.
> The join method between `stars` and `planets` is a hash join, which is marked as `HashJoin_44`. The data access method on `planets` is a `TableFullScan_45`. After the join, the `TopN_26` and `TOPN_19` is to implemented the two order by and limit corespondingly. The final operator `Projection_16` is to implemented the column projection for `t5.name`.

Reviewer 2: ```suggestion
Let's apply the "first child first – recursive descent" rule to the second plan. In the below example begin from the top to bottom, by looking at the `IndexRangeScan_47` (the first child of the tree). For the table `stars`, the optimizer ony need to select the column `name` and `id`, the two columns can be met by the index `name(name)`. So for the table `star`, the root reader is `IndexReader_48`, rather than a `TableReader`. The join method between `stars` and `planets` is a hash join, which is marked as `HashJoin_44`. The data access method on `planets` is a `TableFullScan_45`. After the join, the `TopN_26` and `TOPN_19` is to implemented the two order by and limit corespondingly. The final operator `Projection_16` is to implemented the column projection for `t5.name`.
```

```
MANUAL LINE BREAKS: L389
```

```

## Index Strategy in TiDB

TiDB is a distributed SQL database that completely decouples the SQL layer (TiDB) from the storage layer (TiKV). Unlike traditional databases, TiDB does not have a buffer pool to cache data at the compute node. As a result, the performance of SQL queries and the TiDB cluster is closely tied to the number of key-value (KV) requests that need to be processed.

In TiDB, leveraging indexes effectively is crucial for performance tuning, as it can significantly reduce the number of KV [RPC](/glossary.md#remote-procedure-call-rpc) requests. By minimizing these requests, you can greatly improve query performance and overall system efficiency. Here are some key strategies:

- Avoiding full table scans
- Avoiding sorting
- Skipping row lookups when possible

This section showcases three practical examples that illustrate effective indexing strategies in TiDB, focusing on the use of composite and covered indexes for both filtering and sorting operations.

### SQL Tuning with a Covered Index

> A covered index is designed to include all columns referenced in the filter and select clauses. The query below requires an index lookup of 2597411 rows, taking 46.4 seconds to execute.
> TiDB needs to dispatch 67 cop tasks for the index range scan on logs_idx, identified as `IndexRangeScan_11`, and 301 cop tasks for table access via `TableRowIDScan_12`.
> By utilizing a covered index, the index lookup can be avoided, leading to improved performance.

Reviewer 2: ```suggestion
A covered index is designed to include all columns referenced in the filter and select clauses. The query below requires an index lookup of 2597411 rows, taking 46.4 seconds to execute. TiDB needs to dispatch 67 cop tasks for the index range scan on logs_idx, identified as `IndexRangeScan_11`, and 301 cop tasks for table access via `TableRowIDScan_12`. By utilizing a covered index, the index lookup can be avoided, leading to improved performance.
```

```
MANUAL LINE BREAKS: L526
MANUAL LINE BREAKS: L527
```

|     └─Selection_12            | 5245632.33 | 52863   | cop[tikv] |                                                                                                                                 | tikv_task:{time:80ms, loops:52}          ...|
|       └─IndexRangeScan_11     | 5245632.33 | 52863   | cop[tikv] | table:logs, index:logs_covered(snapshot_id, user_id, status, source_type, target_type, amount)                                  | tikv_task:{time:60ms, loops:52}          ...|
+-------------------------------+------------+---------+-----------+---------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------+
```

### SQL Tuning with a Composite Index Involing Sorting

When optimizing SQL queries, especially those that include an `ORDER BY` clause, it is beneficial to create a composite index that encompasses both the filtering and sorting columns. This approach allows the database engine to efficiently access the required data while maintaining the desired order.

> For instance, consider the following query that retrieves logs based on specific conditions. The execution plan shows a duration of 170ms. TiDB employs the `logs_index` to perform an `IndexRangeScan_20` with the filter `snapshot_id = 459840`. Subsequently, it retrieves all columns from the table, resulting in 5715 rows being streamed back to TiDB after the `IndexLookUp_23`, which then sorts the dataset and returns 1000 rows.
> Note that the `id` column is the primary key, which means it is implicitly included in the `logs_idx` index. However, for `IndexRangeScan_20`, the order is not guaranteed because, after the index prefix column `snapshot_id`, there are two additional columns: `user_id` and `status`. Therefore, the ordering of `id` cannot be assured.

Reviewer 2: ```suggestion
For instance, consider the following query that retrieves logs based on specific conditions. The execution plan shows a duration of 170ms. TiDB employs the `logs_index` to perform an `IndexRangeScan_20` with the filter `snapshot_id = 459840`. Subsequently, it retrieves all columns from the table, resulting in 5715 rows being streamed back to TiDB after the `IndexLookUp_23`, which then sorts the dataset and returns 1000 rows. Note that the `id` column is the primary key, which means it is implicitly included in the `logs_idx` index. However, for `IndexRangeScan_20`, the order is not guaranteed because, after the index prefix column `snapshot_id`, there are two additional columns: `user_id` and `status`. Therefore, the ordering of `id` cannot be assured.
```

```
MANUAL LINE BREAKS: L584
```

+----------------------------------+---------+---------+-----------+----------------------------------------------+----------------------------------------------+----------------------------------------------------+
```

### SQL Tuning with Composite Indexes for Efficient Filtering and Sorting

The original query took 11 minutes and 9 seconds to complete, which is an extremely long execution time for a query that only needs to return 101 rows. This poor performance can be attributed to several factors:

1. Inefficient index usage: The optimizer chose the index on created_at, which resulted in scanning 25,147,450 rows.
2. Large intermediate result set: After applying the date range filter, 12,082,311 rows still needed to be processed.
3. Late filtering: The most selective predicates (mode, user_id, and label_id) were applied after accessing the table, resulting in 16,604 rows.
4. Sorting overhead: The final sort operation on 16,604 rows added additional processing time.

> 

Reviewer 2: ```
sql-tuning-best-practice.md: 643: MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
```

```suggestion
```

| │   └─IndexRangeScan_28        | 503893.42 | 224     | cop[tikv] | table:orders, index:index_orders_new(user_id, mode, id, created_at, label_id)  | tikv_task:{time:0s, loops:3}                        | range:[11111 "production" 1000000000,11111 "production" 1500000000), keep order:true, desc                           | N/A      | N/A  |
| └─TableRowIDScan_29(Probe)     | 101.00    | 101     | cop[tikv] | table:orders                                                                   | time:2.9ms, loops:2, cop_task: {num: 3, max: 2.7...}| keep order:false                                                                                                     | N/A      | N/A  |
+--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
```

### Composite Index Strategy Guidelines

> When creating a composite index, it's essential to follow a specific order for the columns to ensure optimal performance. This order is crucial because it directly impacts how efficiently the index can filter and sort data.
> Here is the recommended order for the columns in the index:

Reviewer 2: ```suggestion
When creating a composite index, it's essential to follow a specific order for the columns to ensure optimal performance. This order is crucial because it directly impacts how efficiently the index can filter and sort data.

Here is the recommended order for the columns in the index:
```

```
MANUAL LINE BREAKS: L711
```

3. Additional filtering columns:
   - Non-equal predicates (!=, <>, IS NOT NULL, etc)
   - Time range conditions on datetime columns
   - Helps reduce row lookups

4. columns in index postfix for select list or aggregate function:
   - Leverage IndexReader
   - Avoid IndexLookup operations

### The Cost of Indexing

While indexes can significantly improve query performance, they also come with costs that should be carefully considered:

1. Performance Impact on Writes:
   - Each additional index slows down write operations (INSERT, UPDATE, DELETE)
   - When data is modified, all affected indexes must be updated
   - This overhead increases with each additional index

2. Resource Consumption:
   - Indexes require additional disk space
   - More memory is needed to cache frequently accessed indexes
   - Backup and recovery operations take longer

3. Write Hotspot Risk:
   - Secondary indexes can create write hotspots, for example, a datetime monotonically increasing index will cause the hotspots on the write of the table
   - Performance can degrade significantly if hotspots occur

Best Practice:

- Only create indexes that provide clear performance benefits

> - Regularly review index usage statistics via [TIDB_INDEX_USAGE](https://docs.pingcap.com/tidb/dev/information-schema-tidb-index-usage)

Reviewer 2: ```suggestion
- Regularly review index usage statistics via [TIDB_INDEX_USAGE](/information-schema/information-schema-tidb-index-usage.md)
```

1. The SQL statement with the highest total latency is the one that takes the longest time to execute out of multiple executions of the same SQL statement. 
2. It can also display the number of times each SQL statement has been executed cumulatively, allowing us to find the SQL statement with the highest execution frequency.
3. Clicking on each SQL statement allows us to delve deeper into the `EXPLAIN ANALYZE` results.

SQL statements are normalized as templates, where literals and bind variables are replaced by '?'. This normalization and sorting process allows you to quickly pinpoint the most resource-intensive queries that may require optimization.
![sql-statements-default](/media/sql-tuning/sql-statements-default.png)

### Slow Queries Panel Default Display

In the TiDB Dashboard, we can find the Slow Query panel, which displays all SQL statements whose execution time exceeds the time threshold set by the system variable tidb_slow_log_threshold (default 300 milliseconds). 

On Slow Queries panel, we can find:

1. The slowest SQL queries.
2. The SQL query that reads the most data from TiKV.
3. The `EXPLAIN ANALYZE` output from drilling down the SQL statement by clicking it.

> 4. Please note that on the Slow Queries panel, we cannot get the frequency of the SQL statement execution. Once the execution elapsed time exceeds tidb_slow_log_threshold for single instance, the query is then listed on the Slow Queries panel.

Reviewer 2: ```suggestion
4. Please note that on the Slow Queries panel, we cannot get the frequency of the SQL statement execution. Once the execution elapsed time exceeds [`tidb_slow_log_threshold`](/tidb-configuration-file.md#tidb_slow_log_threshold) for single instance, the query is then listed on the Slow Queries panel.
```

- The column-level statistics information is more abundant, including histograms, Count-Min Sketch, Top-N (values or indexes with the highest occurrences), distribution and quantity of different values, and the number of null values, and so on.

To ensure the statistics are healthy and representative, you can use the following commands:

1. [`SHOW STATS_META`](sql-statements/sql-statement-show-stats-meta.md): This command provides metadata about table statistics.
2. [`SHOW STATS_HEALTHY`](sql-statements/sql-statement-show-stats-healthy.md): This command shows the health status of table statistics.

For example, you can use:

```sql
SHOW STATS_META WHERE table_name='T2'\G;
```

```
*************************** 1. row ***************************
 Db_name: test
 Table_name: T2
Partition_name:
 Update_time: 2023-05-11 02:16:50
 Modify_count: 20000
 Row_count: 20000
1 row in set (0.03 sec)
```

```sql
SHOW STATS_HEALTHY WHERE table_name='T2'\G;
```

```
*************************** 1. row ***************************
    Db_name: test
  Table_name: T2
Partition_name:
    Healthy: 0
1 row in set (0.00 sec)

> ```

Reviewer 2: The formatting of the output looks wrong, I would expect the `:` to be aligned

3. Gather statistics for specific columns only

It's important to note that after manual collection, subsequent automatic gathering jobs will inherit the new parameters. This means that any customizations you've made during manual collection will be carried forward in future automatic analyses.

Another common scenario is locking table statistics. This is useful when:

1. The statistics on the table are already representative of the data.
2. The table is very large and statistics collection is time-consuming.
3. You want to maintain statistics only during specific time windows.

To lock the statistics for a table, you can use the following statement [`LOCK STATS table_name`](/sql-statements/sql-statement-lock-stats.md).

for more detail about statistics, please refer to [statistics](/statistics.md).

## How TiDB build A Execution Plan

An SQL statement undergoes optimization primarily in the optimizer through three stages:

- Pre-Processing
- Logical Transformation
- Cost-based Optimization

### Pre-Processing

The main actions in the pre-processing stage it to determine if the SQL statement can be executed by using Point_Get or Batch_Point_Get.

> Point_Get or Batch_Point_Get is to get 1 or 0 or many row only by using the TiKV key, the explicit or implicit (`_tidb_rowid`) primary key. For example, when id column is the primary key of a clustered index table, Point_Get is used to get the particular row. If a plan is identified as Point_Get, optimizer will skip the logical transformation and cost-based optimization.

Reviewer 2: ```suggestion
Point_Get or Batch_Point_Get is to get 1 or 0 or many row only by using the TiKV key, the explicit or implicit (`_tidb_rowid`) primary key. For example, when id column is the primary key of a [clustered index](/clustered-indexes.md) table, Point_Get is used to get the particular row. If a plan is identified as Point_Get, optimizer will skip the logical transformation and cost-based optimization.
```

### Cost-Based Optimization

TiDB uses statistics as input to the optimizer to estimate the number of rows processed in each plan step for a SQL statement, and associates a cost with each plan step. The Cost-Based Optimization estimates the cost of each available plan choice, including index accesses and the sequence of table joins, and produces a cost for each available plan. The optimizer then picks the execution plan with the lowest overall cost.

The following figure illustrates the various data access paths and row set operations that cost-based optimization can consider when developing the optimal execution plan. For data access paths, the optimizer determines the most efficient method to retrieve data, whether through an index scan or a table scan, and whether to retrieve the data from row-based TiKV or columnar-based TiFlash storage.

Also, the optimizer need to evaluate operations that manipulate row sets, such as aggregation, join, and sorting. For instance, the aggregation operator may utilize either `HASH_AGG` or `STREAM_AGG`, while the join operator can select from `HASH JOIN`, `MERGE JOIN`, or `INDEX JOIN`. 

> Furthermore, expression and operator push-down to the physical storage engines is part of the physical optimization phase. The physical plan is partitioned and distributed to different components based on the underlying storage engines:
> - Root Task runs at the TiDB layer
> - Cop (Coprocessor) Task runs at the TiKV or TiFlash layer
> - MPP (Massively Parallel Processing) Task runs at the TiFlash layer

Reviewer 2: ```suggestion
Furthermore, expression and operator push-down to the physical storage engines is part of the physical optimization phase. The physical plan is partitioned and distributed to different components based on the underlying storage engines:

- Root Task runs at the TiDB layer
- Cop (Coprocessor) Task runs at the TiKV or TiFlash layer
- MPP (Massively Parallel Processing) Task runs at the TiFlash layer
```

```
sql-tuning-best-practice.md: this file has manual line breaks in the following lines:

MANUAL LINE BREAKS: L282
```

    Db_name: test
  Table_name: T2
Partition_name:
    Healthy: 0
1 row in set (0.00 sec)
```

In TiDB database, there are two ways to collect statistics: automatic collection and manual collection. In most case, the auto collection job works fine. Automatic collection is actually triggered when certain conditions are met for a table, and TiDB will automatically collect statistics. We commonly use three triggering conditions, which are: ratio, start_time and end_time.

- tidb_auto_analyze_ratio: The healthiness trigger
- [`tidb_auto_analyze_start_time`](/system-variables.md#tidb_auto_analyze_start_time) and [`tidb_auto_analyze_end_time`](/system-variables.md#tidb_auto_analyze_end_time): The allowed job window

```sql
SHOW VARIABLES LIKE 'tidb\_auto\_analyze%';
```

```
+-----------------------------------------+-------------+
| Variable_name                           | Value       |
+-----------------------------------------+-------------+
| tidb_auto_analyze_ratio                 | 0.5         |
| tidb_auto_analyze_start_time            | 00:00 +0000 |
| tidb_auto_analyze_end_time              | 23:59 +0000 |
+-----------------------------------------+-------------+
```

> In cases where automatic collection doesn't meet your needs, you can manually collect statistics using the `ANALYZE TABLE table_name` statement. This allows you to:

Reviewer 3: In v8.4, we have enabled the predict analyze. if you miss column's stats, you should run ```analyze table table_name all columns```.

Reviewer 2: @hawkingrei maybe https://docs.pingcap.com/tidb/v8.4/sql-statement-analyze-table should also explain a bit more about the `all columns`?

2. Pre-Process is primarily for Point Get. If it is a Point Get, the following-up optimization processes can be skipped, the next step jumps to the SQL Executor.
3. After confirming that it is not a Point Get, the AST goes to the Logical Transformation, which rewrites the SQL logically based on certain rules.
4. The AST that has gone through Logical Transformation will undergo Cost-Based Optimization.
5. During Cost-Based Optimization, the optimizer considers statistics to determine how to select specific operators and finally generates a physically executable physical execution plan.
6. The generated physical execution plan is sent to the SQL Executor of the TiDB database for execution.
7. Unlike traditional databases, TiDB databases need to push down the execution plan to different TiKV or TiFlash for reading and writing data.

![workflow](/media/sql-tuning/workflow-tiflash.png)

## Optimizer Fundamentals

TiDB uses a cost-based optimizer (CBO) to determine the most efficient execution plan for a SQL statement. This optimizer evaluates different execution strategies and chooses the one with the lowest estimated cost. The cost is influenced by factors such as:

- SQL

> - Schema Design

Author: I think after we put the best practice for the partition table online, we can add a schema design session and link to the partition table document

1. Adjust the sample rate for more accurate or faster analysis
2. Increase the number of top-N values collected
3. Gather statistics for specific columns only or for all columns

It's important to note that after manual collection, subsequent automatic gathering jobs will inherit the new parameters. This means that any customizations you've made during manual collection will be carried forward in future automatic analyses.

Another common scenario is locking table statistics. This is useful when:

1. The statistics on the table are already representative of the data.
2. The table is very large and statistics collection is time-consuming.
3. You want to maintain statistics only during specific time windows.

To lock the statistics for a table, you can use the following statement [`LOCK STATS table_name`](/sql-statements/sql-statement-lock-stats.md).

for more detail about statistics, please refer to [statistics](/statistics.md).

## How TiDB Builds A Execution Plan

An SQL statement undergoes optimization primarily in the optimizer through three stages:

- Pre-Processing
- Logical Transformation
- Cost-based Optimization

### Pre-Processing

> The main actions in the pre-processing stage it to determine if the SQL statement can be executed by using Point_Get or Batch_Point_Get.

Reviewer 2: ```suggestion
The main actions in the pre-processing stage it to determine if the SQL statement can be executed by using `Point_Get` or `Batch_Point_Get`.
```

Maybe link to https://docs.pingcap.com/tidb/stable/explain-indexes#point_get-and-batch_point_get

Another common scenario is locking table statistics. This is useful when:

1. The statistics on the table are already representative of the data.
2. The table is very large and statistics collection is time-consuming.
3. You want to maintain statistics only during specific time windows.

To lock the statistics for a table, you can use the following statement [`LOCK STATS table_name`](/sql-statements/sql-statement-lock-stats.md).

for more detail about statistics, please refer to [statistics](/statistics.md).

## How TiDB Builds A Execution Plan

An SQL statement undergoes optimization primarily in the optimizer through three stages:

- Pre-Processing
- Logical Transformation
- Cost-based Optimization

### Pre-Processing

The main actions in the pre-processing stage it to determine if the SQL statement can be executed by using [`Point_Get`](https://docs.pingcap.com/tidb/stable/explain-indexes#point_get-and-batch_point_get) or [`Batch_Point_Get`](https://docs.pingcap.com/tidb/stable/explain-indexes#point_get-and-batch_point_get).

> `Point_Get` or `Batch_Point_Get` is to get 1 or 0 or many row only by using the TiKV key, the explicit or implicit (`_tidb_rowid`) primary key. For example, when id column is the primary key of a [clustered index](/clustered-indexes.md) table, `Point_Get` is used to get the particular row. If a plan is identified as `Point_Get`, optimizer will skip the logical transformation and cost-based optimization.

Reviewer 4: Or maybe:
```suggestion
`Point_Get` or `Batch_Point_Get` means using a primary or unique key, to directly read from TiKV, by exact key lookup. For example, when id column is the primary key of a [clustered index](/clustered-indexes.md) table, `Point_Get` is used to get the particular row. If a plan is identified as `Point_Get`, optimizer will skip the logical transformation and cost-based optimization, since the exact key read will be the best way to access the row. `Batch_Point_Get` will read a batch of exact key lookups, like `SELECT * FROM t WHERE id IN (4,62,81)`

Example:
```

### Cost-Based Optimization

TiDB uses statistics as input to the optimizer to estimate the number of rows processed in each plan step for a SQL statement, and associates a cost with each plan step. The Cost-Based Optimization estimates the cost of each available plan choice, including index accesses and the sequence of table joins, and produces a cost for each available plan. The optimizer then picks the execution plan with the lowest overall cost.

The following figure illustrates the various data access paths and row set operations that cost-based optimization can consider when developing the optimal execution plan. For data access paths, the optimizer determines the most efficient method to retrieve data, whether through an index scan or a table scan, and whether to retrieve the data from row-based TiKV or columnar-based TiFlash storage.

Also, the optimizer need to evaluate operations that manipulate row sets, such as aggregation, join, and sorting. For instance, the aggregation operator may utilize either `HASH_AGG` or `STREAM_AGG`, while the join operator can select from `HASH JOIN`, `MERGE JOIN`, or `INDEX JOIN`. 

> Furthermore, expression and operator push-down to the physical storage engines is part of the physical optimization phase. The physical plan is partitioned and distributed to different components based on the underlying storage engines:

Reviewer 4: I think it is better skip `partitioned and`, since it is overloading the meaning of partitioning which also can be used for table partitioning.

```suggestion
Furthermore, expression and operator push-down to the physical storage engines is part of the physical optimization phase. The physical plan is  distributed to different components based on the underlying storage engines:
```

- Root Task runs at the TiDB layer
- Cop (Coprocessor) Task runs at the TiKV or TiFlash layer
- MPP (Massively Parallel Processing) Task runs at the TiFlash layer

This distribution of the physical plan allows the different components to collaborate and execute the query efficiently.

![cost-based-optimization](/media/sql-tuning/cost-based-optimization.png)

More Detail for Cost-Based Optimization: [SQL Physical Optimization](/sql-physical-optimization.md).

## Understanding Execution Plans

The execution plan represents the steps TiDB will follow to execute a SQL query. In this section, we will learn how to display and read the execution plan.

### Generating and Displaying Execution Plans

Beside access the execution plan information through TiDB Dashboard, TiDB provides a `EXPLAIN` statement to display the execution plan for a SQL query. Here's an example of using `EXPLAIN`:

- id: Operator name and the step unique identifier
- estRows: Estimated number of rows from the particular step

> - task: Indicates the layer where the operator is executed. For instance, `root` indicates execution at the TiDB layer, whereas `cop[tikv]` indicates execution at the TiKV layer. 

Reviewer 4: Should we also mention the TiFlash tasks here?

| StreamAgg_20             | 1.00        | root         |                   | funcs:count(Column#13)->Column#11                                                                  |
| └─TableReader_21         | 1.00        | root         |                   | data:StreamAgg_9                                                                                   |
|   └─StreamAgg_9          | 1.00        | cop[tikv]    |                   | funcs:count(1)->Column#13                                                                          |
|     └─Selection_19       | 250.00      | cop[tikv]    |                   | ge(trips.start_date, 2017-07-01 00:00:00.000000), le(trips.start_date, 2017-07-01 23:59:59.000000) |
|       └─TableFullScan_18 | 10000.00    | cop[tikv]    | table:trips       | keep order:false, stats:pseudo                                                                     |
+--------------------------+-------------+--------------+-------------------+----------------------------------------------------------------------------------------------------+
5 rows in set (0.00 sec)
```

> Let's apply the "first child first – recursive descent" rule to the second plan. In the below example begin from the top to bottom, by looking at the `IndexRangeScan_47` (the first child of the tree). For the table `stars`, the optimizer ony need to select the column `name` and `id`, the two columns can be met by the index `name(name)`. So for the table `star`, the root reader is `IndexReader_48`, rather than a `TableReader`. The join method between `stars` and `planets` is a hash join, which is marked as `HashJoin_44`. The data access method on `planets` is a `TableFullScan_45`. After the join, the `TopN_26` and `TOPN_19` is to implemented the two order by and limit corespondingly. The final operator `Projection_16` is to implemented the column projection for `t5.name`.

Reviewer 4: Could we have the table structures as references in the end of this page, or a link to them?

+--------------------------------------------+-------------+----------+--------------+----------------+--------------------------------------+
```

### SaaS Arbitrary Filtering Workloads

#### Overview

In SaaS applications, it's common to structure tables with composite primary keys that include tenant identification. Let's look at a typical example where TiFlash can significantly improve query performance.

#### Case Study: Multi-tenant Data Access

Consider a table design with a composite primary key: (tenantId, objectTypeId, objectId). A common query pattern is to:

- Fetch the first N records for a specific tenant and object type, and applying random filters across hundreds or thousands of other columns, Making it impractical to create indexes for every possible filter combination. There is a potential sort operator after the filter as well.
- Get the total count of records matching these criterias

#### Performance Comparison

- TiKV Plan: When executing this query on TiKV storage engine, it takes 2 minutes 38.6 seconds as it requires a table range scan operation, and cannot utilize indexes effectively for dynamic filtering conditions across hundreds of columns.

> - TiFlash Plan: The same query on TiFlash MPP engine takes only 3.44 seconds - almost 46 times faster. Since data in TiFlash is sorted by primary key, queries filtered by the primary key's prefix will also use a TableRangeScan instead of a full table scan.

Reviewer 5: It's wrong. Both plans don't contain a full table scan.

From the limited information, I think there are 2 main reasons for the performance difference:
1. Data is spread over 5000+ regions for tikv (which implies 5000+ cop tasks), but tiflash only uses 2 mpp tasks. Besides the RPC overhead, it also means the mpp plan provides a better aggregation effect in the 1st stage aggregation and less aggregation work to do.
2. TiFlash scans and aggregates much faster than TiKV.

Reviewer 4: Would it not be 5000+ regions in TiFlash as well, but for both TiKV and TiFlash it can use the `TenantID` to only use matching regions? (like range partitioning, but directly on the clustered index by regions). But of course RPC overhead may come in to play as well.

Reviewer 5: Yes. I'm not familiar with the implementation of TiFlash, but I believe the data is not stored by regions in TiFlash, and it also doesn't need to split the mpp task according to regions in TiFlash. There only needs to be 1 mpp task sent to 1 TiFlash node for each table, and it's executed as a whole.

| ├─Limit_20                     | 100.00    | 100     | root      |                       | time:23ms, loops:2                                  |
| │ └─TableReader_25             | 100.00    | 100     | root      |                       | time:23ms, loops:1, cop_task: {num: 1, max: 22.8...}|
| │   └─Limit_24                 | 100.00    | 100     | cop[tikv] |                       | tikv_task:{time:21ms, loops:3}, scan_detail: {...}  |
| │     └─TableRangeScan_22      | 100.00    | 100     | cop[tikv] | table:usertable       | tikv_task:{time:21ms, loops:3}                      |
| └─Projection_26                | 1.00      | 1       | root      |                       | time:2m38.6s, loops:2, Concurrency:OFF              |
|   └─HashAgg_34                 | 1.00      | 1       | root      |                       | time:2m38.6s, loops:2, partial_worker:{...}, fin.. .|
|     └─TableReader_35           | 1.00      | 5121    | root      |                       | time:2m38.6s, loops:7, cop_task: {num: 5121, max:...|
|       └─HashAgg_27             | 1.00      | 5121    | cop[tikv] |                       | tikv_task:{proc max:0s, min:0s, avg: 462.8ms, p...} |

> |         └─TableRangeScan_32    | 10000000  | 10000000| cop[tikv] | table:usertable       | tikv_task:{proc max:0s, min:0s, avg: 460.5ms, p...} |

Reviewer 5: The execution info here seems abnormal.

|     └─Selection_83                    | 984.56   | 1998    | cop[tikv] |                                                                                        | tikv_task:{pro...| N/A      | N/A  |
|       └─TableRangeScan_82             | 1000.00  | 2048    | cop[tikv] | table:orders                                                                           | tikv_task:{pro...| N/A      | N/A  |
+---------------------------------------+----------+---------+-----------+----------------------------------------------------------------------------------------+---------------...+----------+------+
```

More Detail for understanding execution plans : [TiDB Query Execution Plan Overview](/explain-overview.md) and [EXPLAIN Walkthrough](/explain-walkthrough.md).

## Index Strategy in TiDB

TiDB is a distributed SQL database that completely decouples the SQL layer (TiDB) from the storage layer (TiKV). Unlike traditional databases, TiDB does not have a buffer pool to cache data at the compute node. As a result, the performance of SQL queries and the TiDB cluster is closely tied to the number of key-value (KV) requests that need to be processed.

> In TiDB, leveraging indexes effectively is crucial for performance tuning, as it can significantly reduce the number of KV [RPC](/glossary.md#remote-procedure-call-rpc) requests. By minimizing these requests, you can greatly improve query performance and overall system efficiency. Here are some key strategies:

Reviewer 4: Btw, do we also mean that `cop task` is the same as RPC call in this context? Then better to connect and explain it here, since just below will mention `cop task`.

- Avoiding full table scans
- Avoiding sorting

> - Skipping row lookups when possible

Reviewer 4: ```suggestion
- Skipping row lookups when possible by using covering index or not request non-needed columns
```

    `logs`.`source_type` IN ('user', 'payment')
    OR `logs`.`source_type` IN (
      'bank_account',
    )
    AND `logs`.`target_type` IN ('bank_account')
  );
```

```
+-------------------------------+------------+---------+-----------+--------------------------------------------------------------------------+------------------------------------------------------------+
| id                            | estRows    | actRows | task      | access object                                                            | execution info                                             | 
+-------------------------------+------------+---------+-----------+--------------------------------------------------------------------------+------------------------------------------------------------+
| HashAgg_18                   | 1.00       | 2571625.22 | 1       | root      |                                                              | time:46.4s, loops:2, partial_worker:{wall_time:46.37,   ...|
| └─IndexLookUp_19             | 1.00       | 2570096.68 | 301     | root      |                                                              | time:46.4s, loops:2, index_task: {total_time: 45.8s,    ...|

> |   ├─IndexRangeScan_11(Build) | 1309.50    | 317033.98  | 2597411 | cop[tikv] | table:logs, index:logs_idx(snapshot_id, user_id, status)     | time:228ms, loops:2547, cop_task: {num: 67, max: 2.17s, ...|

Reviewer 4: Should we run `ANALYZE TABLE logs` first, or is this difference between `estRows` and `actRows` correct?

|     └─Selection_12            | 5245632.33 | 52863   | cop[tikv] |                                                                                                                                 | tikv_task:{time:80ms, loops:52}          ...|
|       └─IndexRangeScan_11     | 5245632.33 | 52863   | cop[tikv] | table:logs, index:logs_covered(snapshot_id, user_id, status, source_type, target_type, amount)                                  | tikv_task:{time:60ms, loops:52}          ...|
+-------------------------------+------------+---------+-----------+---------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------+
```

### SQL Tuning with a Composite Index Involving Sorting

When optimizing SQL queries, especially those that include an `ORDER BY` clause, it is beneficial to create a composite index that encompasses both the filtering and sorting columns. This approach allows the database engine to efficiently access the required data while maintaining the desired order.

> For instance, consider the following query that retrieves logs based on specific conditions. The execution plan shows a duration of 170ms. TiDB employs the `logs_index` to perform an `IndexRangeScan_20` with the filter `snapshot_id = 459840`. Subsequently, it retrieves all columns from the table, resulting in 5715 rows being streamed back to TiDB after the `IndexLookUp_23`, which then sorts the dataset and returns 1000 rows. Note that the `id` column is the primary key, which means it is implicitly included in the `logs_idx` index. However, for `IndexRangeScan_20`, the order is not guaranteed because, after the index prefix column `snapshot_id`, there are two additional columns: `user_id` and `status`. Therefore, the ordering of `id` cannot be assured.

Reviewer 4: Do you mean the `logs_idx` instead of `logs_index`? Also where is the table/index definitions?
Secondly depending on how often this specific query is run, I don't think it is too bad to sort 5k rows :) 170->96ms is good, but not amazing ;)

Also are all columns needed? If not maybe combine with yet another covering index?

| Limit_14                         | 17.59   | 1000    | root      |                                              | time:96.1ms, loops:2, RU:92.300155           | offset:0, count:1000                               |
| └─IndexLookUp_35                 | 17.59   | 1000    | root      |                                              | time:96.1ms, loops:1, index_task:         ...|                                                    |
|   ├─IndexRangeScan_33(Build)     | 17.59   | 5715    | cop[tikv] | table:logs, index:logs_new(snapshot_id, id)  | time:7.25ms, loops:8, cop_task: {num: 3,  ...| range:(459840 998464,459840 +inf], keep order:true |
|   └─TableRowIDScan_34(Probe)     | 17.59   | 5715    | cop[tikv] | table:logs                                   | time:232.9ms, loops:9, cop_task: {num: 3, ...| keep order:false                                   |
+----------------------------------+---------+---------+-----------+----------------------------------------------+----------------------------------------------+----------------------------------------------------+
```

### SQL Tuning with Composite Indexes for Efficient Filtering and Sorting

> The original query took 11 minutes and 9 seconds to complete, which is an extremely long execution time for a query that only needs to return 101 rows. This poor performance can be attributed to several factors:

Reviewer 4: ```suggestion
The original query below took 11 minutes and 9 seconds to complete, which is an extremely long execution time for a query that only needs to return 101 rows. This poor performance can be attributed to several factors:
```

> 1. Inefficient index usage: The optimizer chose the index on created_at, which resulted in scanning 25,147,450 rows.

Reviewer 4: ```suggestion
1. Inefficient index usage: The optimizer chose the index on `created_at`, which resulted in scanning 25,147,450 rows.
```

2. Large intermediate result set: After applying the date range filter, 12,082,311 rows still needed to be processed.
3. Late filtering: The most selective predicates (mode, user_id, and label_id) were applied after accessing the table, resulting in 16,604 rows.
4. Sorting overhead: The final sort operation on 16,604 rows added additional processing time.

Here is the query pattern

```sql
SELECT `orders`.*
FROM `orders`
WHERE 
    `orders`.`mode` = 'production'
    AND `orders`.`user_id` = 11111
    AND orders.label_id IS NOT NULL
    AND orders.created_at >= '2024-04-07 18:07:52'
    AND orders.created_at <= '2024-05-11 18:07:52'

>     AND id >= 1000000000
>     AND id <= 2000000000
>     AND orders.id < 1500000000

Reviewer 4: ```suggestion
    AND orders.id >= 1000000000
    AND orders.id < 1500000000
```

|   └─Selection_38(Probe)        | 173.83    | 16604   | cop[tikv] |                                                                                | time:54m46.2s, loops:651, cop_task: {num: 1076, ...}| eq(orders.mode, "production"), eq(orders.user_id, 11111), not(isnull(orders.label_id))                               | N/A      | N/A  |
|     └─TableRowIDScan_36        | 8296.70   | 12082311| cop[tikv] | table:orders                                                                   | tikv_task:{proc max:44.8s, min:0s, avg: 3.33s, p...}| keep order:false                                                                                                     | N/A      | N/A  |
+--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
```

Performance Improvement with the New Index:

> After creating the new index (new_idx on orders(user_id, mode, id, created_at, label_id)), the query performance improved dramatically. The execution time reduced from 11 minutes and 9 seconds to just 5.3 milliseconds, which is a staggering improvement of over 126,000 times faster. This massive improvement can be explained by:

Reviewer 4: ```suggestion
After creating the new index `idx_composite` on orders(user_id, mode, id, created_at, label_id), the query performance improved dramatically. The execution time reduced from 11 minutes and 9 seconds to just 5.3 milliseconds, which is a staggering improvement of over 126,000 times faster. This massive improvement can be explained by:
```

1. Efficient index usage: The new index allows for an index range scan on user_id, mode, and id, which are the most selective predicates. This drastically reduces the number of rows scanned from millions to just 224.
2. Index-only sort: The 'keep order:true' in the plan indicates that the index structure is used for sorting, avoiding a separate sort operation.
3. Early filtering: The most selective predicates are applied first, quickly reducing the result set to 224 rows before further filtering.
4. Limit push-down: The LIMIT clause is pushed down to the index scan, allowing early termination of the scan once 101 rows are found.

> This case demonstrates the profound impact that a well-designed index can have on query performance. By aligning the index structure with the query's predicates, sort order, and required columns, we achieved a performance improvement of over five orders of magnitude.

Reviewer 4: 'and required columns' you mean 'filtered columns', which is the predicates? Since the query was 'SELECT * ...'

+--------------------------------+-----------+---------+-----------+--------------------------------------------------------------------------------+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+----------+------+
```

### Composite Index Strategy Guidelines

When creating a composite index, it's essential to follow a specific order for the columns to ensure optimal performance. This order is crucial because it directly impacts how efficiently the index can filter and sort data.

Here is the recommended order for the columns in the index:

1. Equal predicates for index prefix (columns accessed directly):
   - Equal conditions 
   - IS NULL conditions

2. Columns after index prefix for sorting:
   - Leverage index to preprocess sorting
   - Ensure sort and limit pushdown to TiKV
   - Maintain sorted order

> 3. Additional filtering columns:
>    - Non-equal predicates (`!=`, `<>`, `IS NOT NULL`, etc)
>    - Time range conditions on datetime columns
>    - Helps reduce row lookups

Reviewer 4: ```suggestion
3. Additional filtering columns, to reduce row lookups:
   - Non-equal predicates (`!=`, `<>`, `IS NOT NULL`, etc)
   - Time range conditions on datetime columns
```

4. columns in index postfix for select list or aggregate function:
   - Leverage IndexReader
   - Avoid IndexLookup operations

### The Cost of Indexing

While indexes can significantly improve query performance, they also come with costs that should be carefully considered:

1. Performance Impact on Writes:

>    - Each additional index slows down write operations (INSERT, UPDATE, DELETE)

Reviewer 4: ```suggestion
   - First non-clustered index decreases the chance of single phase commit optimization.
   - Each additional index slows down write operations (INSERT, UPDATE, DELETE)
```

   - When data is modified, all affected indexes must be updated
   - This overhead increases with each additional index

2. Resource Consumption:
   - Indexes require additional disk space
   - More memory is needed to cache frequently accessed indexes
   - Backup and recovery operations take longer

3. Write Hotspot Risk:
   - Secondary indexes can create write hotspots, for example, a datetime monotonically increasing index will cause the hotspots on the write of the table
   - Performance can degrade significantly if hotspots occur

Best Practice:

- Only create indexes that provide clear performance benefits
- Regularly review index usage statistics via [TIDB_INDEX_USAGE](/information-schema/information-schema-tidb-index-usage.md)
- Consider the write/read ratio of your workload when designing indexes

## When to Use TiFlash: A Simple Guide

This section explains the best scenarios for using TiFlash in TiDB. TiFlash is optimized for analytical queries that involve complex calculations, aggregations, and large dataset scans, thanks to its columnar storage format. Here’s when to consider TiFlash:

> 

Reviewer 4: ```suggestion

- Large single column scans and aggregations: The core of TiFlash is that each column is stored separately, gaining better compression rates, resulting in less data to read and process when a only a single column is needed.
```

The plan for executing the same query on the TiFlash MPP (Massively Parallel Processing) takes only 1.41 seconds, which is 15 times faster than the TiKV plan.

In the TiFlash MPP plan:

- The optimizer recognizes that both the order_line and item tables have TiFlash replicas available.
- As a result, the optimizer is able to push down the entire query execution, including the table scans, hash join, column projection, and aggregation, to the TiFlash MPP layer.
- Performing the entire query execution on the TiFlash MPP layer, which is optimized for analytical workloads, leads to the significant performance improvement compared to the TiKV plan.

The key difference between the two plans is that the TiFlash MPP plan can leverage the distributed processing capabilities of the TiFlash MPP layer to execute the entire query more efficiently, while the TiKV plan requires more coordination between the TiDB and TiKV components.

```sql
select 100.00 * sum(case when i_data like 'PR%' then ol_amount else 0 end) / (1+sum(ol_amount)) as promo_revenue
from	order_line, item

> where	ol_i_id = i_id and ol_delivery_d >= '2007-01-02 00:00:00.000000' and ol_delivery_d < '2030-01-02 00:00:00.000000';

Reviewer 4: Are the columns really named like this? Like `ol_i_id` for `order_line`.`id`, and `i_id` for `item`.`id`?

---
title: A Practical Guide for SQL Tuning
summary: Learn how to optimize SQL queries for better performance.
---

# A Practical Guide for SQL Tuning

> This guide is designed for TiDB SQL tuning beginners. It focuses on:

Reviewer 6: ```suggestion
This guide is designed for TiDB SQL tuning beginners. It focuses on the following key principles:
```

- Schema design
- Statistics, including:
    - Table statistics
    - Index statistics
    - Column statistics

Based on these inputs, the cost model produces an execution plan that details how the system will execute the SQL, including:

- Access method
- Join method
- Join order

The effectiveness of the optimizer depends on the quality of the information it receives. To achieve optimal performance, ensure that statistics are up to date and indexes are well-designed.

### Statistics management

Statistics are essential for the TiDB optimizer. TiDB uses statistics as the input of optimizer to estimate the number of rows processed in each step of a SQL execution plan.

Statistics are divided into two levels:

- **Table-level statistics**: include the total number of rows in the table and the number of rows modified since the last statistics collection.
- **Index/column-level statistics**: include detailed information such as histograms, Count-Min Sketch, Top-N (values or indexes with the highest occurrences), distribution and quantity of different values, and the number of NULL values.

To check the accuracy and health of your statistics, you can use the following SQL statements:

> - [`SHOW STATS_META`](sql-statements/sql-statement-show-stats-meta.md): shows metadata about table statistics.

Reviewer 6: ```suggestion
- [`SHOW STATS_META`](/sql-statements/sql-statement-show-stats-meta.md): shows metadata about table statistics.
```

2. Enhancing data access methods:
    - Use covering indexes to satisfy queries directly from the index, avoiding full table scans.
    - Implement partitioning strategies to limit data scans to relevant partitions.

Examples:

- Create indexes on frequently queried columns to significantly reduce resource usage, particularly for queries that access small portions of the table.
- Use index-only scans for queries that return a limited number of sorted results to avoid full table scans and sorting operations.

### Balance workload distribution

In a distributed architecture like TiDB, balancing workloads across TiKV nodes is essential for optimal performance. To identify and resolve read and write hotspots, see [Troubleshoot hotspot issues](/troubleshoot-hot-spot-issues.md#optimization-of-small-table-hotspots).

By implementing these strategies, you can ensure that your TiDB cluster efficiently utilizes all available resources and avoids bottlenecks caused by uneven workload distribution or serialization on individual TiKV nodes.

## Identify high-load SQL

> The most efficient way to identify resource-intensive SQL statements is by using [TiDB Dashboard](dashboard/dashboard-overview.md). You can also use other tools, such as views and logs, to identify high-load SQL statements.

Reviewer: ```suggestion
The most efficient way to identify resource-intensive SQL statements is by using [TiDB Dashboard](/dashboard/dashboard-overview.md). You can also use other tools, such as views and logs, to identify high-load SQL statements.
```

## SQL tuning guide

This guide provides practical advice for beginners on optimizing SQL queries in TiDB. By following these best practices, you can improve query performance and streamline SQL tuning. This guide covers the following topics:

- Understand query processing
    - Query processing workflow
    - Optimizer fundamentals
    - Statistics management
- Understand execution plans
    - How TiDB builds an execution plan
    - Generate and display execution plans
    - Read execution plans: first child first
    - Identify and understand bottlenecks in execution plans
- Index strategies in TiDB
    - Composite index strategy guidelines
    - The cost of indexing
    - SQL tuning with a covering index
    - SQL tuning with a composite index involving sorting
    - SQL tuning with a composite index for efficient filtering and sorting
- When to use TiFlash

### Query processing workflow

When a client sends a SQL statement to TiDB, the statement passes through the protocol layer of the TiDB server. This layer manages the connection between the TiDB server and the client, receives SQL statements, and returns data to the client.

> In the following figure, to the right of the protocol layer is the optimizer of the TiDB server, which processes SQL statements as follows:

Reviewer: ```suggestion

![workflow](/media/sql-tuning/workflow-tiflash.png)

In the preceding figure, to the right of the protocol layer is the optimizer of the TiDB server, which processes SQL statements as follows:
```

1. The SQL statement arrives at the SQL optimizer through the protocol layer and is parsed into an abstract syntax tree (AST).

> 2. TiDB identifies whether it is a [Point Get](/explain-indexes#point_get-and-batch_point_get) statement, which involves a simple one-table lookup through a primary or unique key, such as `SELECT * FROM t WHERE pk_col = 1` or `SELECT * FROM t WHERE uk_col IN (1,2,3)`. For `Point Get` statements, TiDB skips subsequent optimization steps and proceeds directly to execution in the SQL executor.

Reviewer: ```suggestion
2. TiDB identifies whether it is a [Point Get](/explain-indexes.md#point_get-and-batch_point_get) statement, which involves a simple one-table lookup through a primary or unique key, such as `SELECT * FROM t WHERE pk_col = 1` or `SELECT * FROM t WHERE uk_col IN (1,2,3)`. For `Point Get` statements, TiDB skips subsequent optimization steps and proceeds directly to execution in the SQL executor.
```

3. If the query is not a `Point Get`, the AST undergoes logical transformation, where TiDB rewrites the SQL logically based on specific rules.
4. After logical transformation, TiDB processes the AST through cost-based optimization.
5. During cost-based optimization, the optimizer uses statistics to select appropriate operators and generates a physical execution plan.
6. The generated physical execution plan is sent to the SQL executor of the TiDB node for execution.
7. Unlike traditional single-node databases, TiDB pushes down operators or coprocessors to TiKV and/or TiFlash nodes containing the data. This approach processes parts of the execution plan where the data is stored, efficiently utilizing the distributed architecture, using resources in parallel, and reducing network data transfer. The TiDB node executor then assembles the final result and returns it to the client.

> 
> ![workflow](/media/sql-tuning/workflow-tiflash.png)

Reviewer: ```suggestion
```

Examples:

- Create indexes on frequently queried columns to significantly reduce resource usage, particularly for queries that access small portions of the table.
- Use index-only scans for queries that return a limited number of sorted results to avoid full table scans and sorting operations.

### Balance workload distribution

In a distributed architecture like TiDB, balancing workloads across TiKV nodes is essential for optimal performance. To identify and resolve read and write hotspots, see [Troubleshoot hotspot issues](/troubleshoot-hot-spot-issues.md#optimization-of-small-table-hotspots).

By implementing these strategies, you can ensure that your TiDB cluster efficiently utilizes all available resources and avoids bottlenecks caused by uneven workload distribution or serialization on individual TiKV nodes.

## Identify high-load SQL

The most efficient way to identify resource-intensive SQL statements is by using [TiDB Dashboard](/dashboard/dashboard-overview.md). You can also use other tools, such as views and logs, to identify high-load SQL statements.

### Monitor SQL statements using TiDB Dashboard

#### SQL Statements page

> In [TiDB Dashboard](dashboard/dashboard-overview.md), navigate to the [**SQL Statements** page](/dashboard/dashboard-statement-list.md) to identify the following:

Reviewer 6: ```suggestion
In [TiDB Dashboard](/dashboard/dashboard-overview.md), navigate to the [**SQL Statements** page](/dashboard/dashboard-statement-list.md) to identify the following:
```

- The SQL statement with the highest total latency, which is the statement that takes the longest time to execute across multiple executions.
- The number of times each SQL statement has been executed, which helps identify statements with the highest execution frequency.
- Execution details of each SQL statement by clicking it to view `EXPLAIN ANALYZE` results.

TiDB normalizes SQL statements into templates by replacing literals and bind variables with `?`. This normalization and sorting process helps you quickly identify the most resource-intensive queries that might require optimization.

![sql-statements-default](/media/sql-tuning/sql-statements-default.png)

#### Slow Queries page

> In [TiDB Dashboard](dashboard/dashboard-overview.md), navigate to the [**Slow Queries** page](/dashboard/dashboard-slow-query.md) to find the following:

Reviewer 6: ```suggestion
In [TiDB Dashboard](/dashboard/dashboard-overview.md), navigate to the [**Slow Queries** page](/dashboard/dashboard-slow-query.md) to find the following:
```

```sql
PLAN REPLAYER DUMP EXPLAIN [ANALYZE] [WITH STATS AS OF TIMESTAMP expression] sql-statement;
```

Use `EXPLAIN ANALYZE` whenever possible, as it provides both the execution plan and actual performance metrics for more accurate insights into query performance.

## SQL tuning guide

This guide provides practical advice for beginners on optimizing SQL queries in TiDB. By following these best practices, you can improve query performance and streamline SQL tuning. This guide covers the following topics:

- [Understand query processing](#understand-query-processing)
    - [Query processing workflow](#query-processing-workflow)
    - [Optimizer fundamentals](#optimizer-fundamentals)
    - [Statistics management](#statistics-management)
- [Understand execution plans](#understand-execution-plans)
    - [How TiDB builds an execution plan](#how-tidb-builds-an-execution-plan)
    - [Generate and display execution plans](#generate-and-display-execution-plans)
    - [Read execution plans: first child first](#read-execution-plans-first-child-first)

>     - [Identify and understand bottlenecks in execution plans](#identify-and-understand-bottlenecks-in-execution-plans)

Reviewer 6: ```suggestion
    - [Identify bottlenecks in execution plans](#identify-bottlenecks-in-execution-plans)
```

> - [Index strategies in TiDB](#index-strategies-in-tidb)

Reviewer 6: ```suggestion
- [Index strategy in TiDB](#index-strategy-in-tidb)
```

    - [Composite index strategy guidelines](#composite-index-strategy-guidelines)
    - [The cost of indexing](#the-cost-of-indexing)
    - [SQL tuning with a covering index](#sql-tuning-with-a-covering-index)
    - [SQL tuning with a composite index involving sorting](#sql-tuning-with-a-composite-index-involving-sorting)

>     - [SQL tuning with a composite index for efficient filtering and sorting](#sql-tuning-with-a-composite-index-for-efficient-filtering-and-sorting)

Reviewer 6: ```suggestion
    - [SQL tuning with composite indexes for efficient filtering and sorting]#sql-tuning-with-composite-indexes-for-efficient-filtering-and-sorting
```

Note that after manual collection, subsequent automatic collection jobs inherit the new settings. This means that any customizations made during manual collection will apply to future automatic analyses.

Locking table statistics is useful in the following scenarios:

- The statistics on the table already represent the data well.
- The table is very large, and statistics collection is time-consuming.
- You want to maintain statistics only during specific time windows.

To lock statistics for a table, you can use the [`LOCK STATS table_name`](/sql-statements/sql-statement-lock-stats.md) statement.

For more information, see [Statistics](/statistics.md).

### Understand execution plans

An execution plan details the steps that TiDB will follow to execute a SQL query. This section explains how TiDB builds an execution plan and how to generate, display, and interpret execution plans.

#### How TiDB builds an execution plan

A SQL statement undergoes three main optimization stages in the TiDB optimizer:

1. [Pre-processing](#1-pre-processing)
2. [Logical transformation](#2-logical-transformation)
3. [Cost-based optimization](#3-cost-based-optimization)

##### 1. Pre-processing

> During pre-processing, TiDB determines whether the SQL statement can be executed using [`Point_Get`](/explain-indexes#point_get-and-batch_point_get) or [`Batch_Point_Get`](/explain-indexes#point_get-and-batch_point_get). These operations use a primary or unique key to read directly from TiKV through an exact key lookup. If a plan qualifies for `Point_Get` or `Batch_Point_Get`, the optimizer skips the logical transformation and cost-based optimization steps because direct key lookup is the most efficient way to access the row.

Reviewer 6: ```suggestion
During pre-processing, TiDB determines whether the SQL statement can be executed using [`Point_Get`](/explain-indexes.md#point_get-and-batch_point_get) or [`Batch_Point_Get`](/explain-indexes.md#point_get-and-batch_point_get). These operations use a primary or unique key to read directly from TiKV through an exact key lookup. If a plan qualifies for `Point_Get` or `Batch_Point_Get`, the optimizer skips the logical transformation and cost-based optimization steps because direct key lookup is the most efficient way to access the row.
```
